# Preparation


Variables:

```shell
DIR_REPOSITORY=/lizardfs/guarracino/pggb-paper
```

## Primates

Download:

```shell
mkdir -p $DIR_BASE/assemblies

# T2T.primates_assemblies.urls.txt refers to assemblies in https://genomeark.github.io/t2t-draft-assembly/, but they were taken from https://genomeark.s3.amazonaws.com/index.html?prefix=species/ (11 January 2023)
# https://docs.google.com/spreadsheets/d/1bWZ1SjCn6I34QMqXeg-zwzK7D-EDizw1GQCkQ_Ily6E/edit#gid=0
sbatch -c 1 --wrap "cd $DIR_BASE/assemblies; (echo https://hgdownload.soe.ucsc.edu/goldenPath/hs1/bigZips/hs1.fa.gz; cat ../data/T2T.primates_assemblies.urls.txt) | parallel -j 8 'wget -q {} && echo got {}'"
mv hs1.fa.gz chm13v2.fasta.gz

# Apply PanSN-spec
ls *fasta.gz | while read f; do
  echo $f
  prefix=$(echo $f | cut -f 1 -d '.')

  $RUN_FASTIX -p "${prefix}#" <(zcat $f ) | sed -e 's/haplotype//g' -e 's/-/#/g' -e 's/unassigned/U/g' | bgzip -c -@ 48 > $prefix.fa.gz
  samtools faidx $prefix.fa.gz
done
```

Map contigs against the references:

```shell
mkdir -p $DIR_BASE/partitioning
cd $DIR_BASE/partitioning

REFS=/lizardfs/guarracino/chromosome_communities/assemblies/chm13v2+grch38masked.fa.gz

ls $DIR_BASE/assemblies/*.fa.gz | while read FASTA; do
  SPECIES=$(basename $FASTA .fa.gz | cut -f 1,2 -d '.');
  echo $SPECIES
  
  PAF=$DIR_BASE/partitioning/$SPECIES.vs.ref.p90.paf
  sbatch -p workers -c 20 --wrap "$RUN_WFMASH -t 20 -m -N -s 50k -l 150k -p 90 -H 0.001 $REFS $FASTA > $PAF"
done
```

Collect unmapped contigs and remap them in split mode:

```shell
REFS=/lizardfs/guarracino/chromosome_communities/assemblies/chm13v2+grch38masked.fa.gz

(ls $DIR_BASE/assemblies/*.fa.gz | grep 'chm13\|primates' -v) | while read FASTA; do
  SPECIES=$(basename $FASTA .fa.gz | cut -f 1,2 -d '.');
  echo $SPECIES
  
  UNALIGNED=$DIR_BASE/partitioning/$SPECIES.unaligned
  
  PAF=$DIR_BASE/partitioning/$SPECIES.vs.ref.p90.paf
  comm -23 <(cut -f 1 $FASTA.fai | sort) <(cut -f 1 $PAF | sort) > $UNALIGNED.txt
  if [[ $(wc -l $UNALIGNED.txt | cut -f 1 -d ' ' ) != 0 ]];
  then 
    samtools faidx $FASTA $(tr '\n' ' ' < $UNALIGNED.txt) > $UNALIGNED.fa
    samtools faidx $UNALIGNED.fa
    sbatch -p workers -c 20 --wrap "$RUN_WFMASH -t 20 -m -s 50k -l 150k -p 90 -H 0.001 $REFS $UNALIGNED.fa > $UNALIGNED.split.vs.ref.p90.paf"
  fi
done
```

Collect our best mapping for each of our attempted split rescues:

```shell
ls *.unaligned.split.vs.ref.p90.paf | while read PAF; do
  cat $PAF | awk -v OFS='\t' '{ print $1,$11,$0 }' | sort -n -r -k 1,2 | \
    awk -v OFS='\t' '$1 != last { print($3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15); last = $1; }'
done > rescues.paf
```

Collect partitioned contigs:

```shell
# Take only haplotype-assigned contigs
( seq 6 7 8 ) | while read i; do cat *paf | grep -P -e "[chm13|grch38]#chr$i\t" | grep '#U#' -v | cut -f 1 | sort | uniq | awk '{print($0"$")}' > chr$i.contigs.txt; done

( seq 6 7 8  ) | while read i; do
    echo chr$i

    rm chr$i.fa*
    (ls $DIR_BASE/assemblies/*.fa.gz | grep 'chm13\|primates' -v) | while read FASTA; do
      echo $FASTA

      samtools faidx $FASTA $(cut -f 1 $FASTA.fai | grep -f chr$i.contigs.txt) >> chr$i.fa
    done
done

REFS=/lizardfs/guarracino/chromosome_communities/assemblies/chm13v2+grch38masked.fa.gz

# Only chr 6 and 7 are available from primates4
( seq 6 7 ) | while read i; do
    echo chr$i
    # 4 haplotypes (from primates4) + 12 haplotypes (6 diploid assemblies) + 1 haplotype (chm13#chr$i)
    cat \
      chr$i.fa \
      <(zcat /lizardfs/guarracino/pggb-paper/assemblies/primates4.chr$i.fa.gz) \
      <(samtools faidx $REFS $(grep chm13#chr$i $REFS.fai | cut -f 1)) \
      > primates17.chr$i.fa
    bgzip -@ 48 primates17.chr$i.fa
    samtools faidx primates17.chr$i.fa.gz
done

( seq 8 8 ) | while read i; do
    echo chr$i
    # 12 haplotypes (6 diploid assemblies) + 2 haplotype (chm13#chr$i and grch38#chr$i)
    cat \
      chr$i.fa \
      <(samtools faidx $REFS $(grep hr$i $REFS.fai | cut -f 1)) \
      > primates14.chr$i.fa
    bgzip -@ 48 primates14.chr$i.fa
    samtools faidx primates14.chr$i.fa.gz
done
```

## Data compression

```shell
cd $DIR_REPOSITORY/assemblies

ls *fa | while read FASTA; do
    echo $FASTA
    zstd -12 $FASTA
done

# "zstd -d file.zst" to decompress
```

## Tools

```shell
mkdir -p ~/tools $$ cd ~/tools

git clone --recursive https://github.com/waveygang/wfmash
cd wfmash
git checkout master && git pull && git submodule update --init --recursive
git checkout 191afe12042962d3c0d5c62936528561753b3da0
cmake -H. -Bbuild && cmake --build build -- -j 48
cp build/bin/wfmash build/bin/wfmash-191afe12042962d3c0d5c62936528561753b3da0
cd ..

git clone --recursive https://github.com/ekg/seqwish.git
cd seqwish
git checkout master && git pull && git submodule update --init --recursive
git checkout d9e7ab59e73258f57875f2a060437735a460475e
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/seqwish bin/seqwish-d9e7ab59e73258f57875f2a060437735a460475e
cd ..

git clone --recursive https://github.com/pangenome/smoothxg.git
cd smoothxg
git checkout master && git pull && git submodule update --init --recursive
git checkout 06bbf3510b095f1f2cb94f580845bfd5097f3047
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/smoothxg bin/smoothxg-06bbf3510b095f1f2cb94f580845bfd5097f3047
cd ..

git clone --recursive https://github.com/pangenome/odgi.git
cd odgi
git checkout master && git pull && git submodule update --init --recursive
git checkout bfae0b3c70a0f044d7ce8510b7a45f59d5138626
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/odgi bin/odgi-bfae0b3c70a0f044d7ce8510b7a45f59d5138626
cd ..

git clone --recursive https://github.com/pangenome/pggb.git
cd pggb
git checkout master && git pull && git submodule update --init --recursive
git checkout b3c55023313c28d89281d2e0ca0f10d25685d4e7
cp pggb pggb-x
sed 's,"$fmt" wfmash,"$fmt" ~/tools/wfmash/build/bin/wfmash-191afe12042962d3c0d5c62936528561753b3da0,g' pggb-x -i
sed 's,$(wfmash,$(~/tools/wfmash/build/bin/wfmash-191afe12042962d3c0d5c62936528561753b3da0,g' pggb-x -i
sed 's,"$fmt" seqwish,"$fmt" ~/tools/seqwish/bin/seqwish-f362f6f5ea89dbb6a0072a8b8ba215e663301d33,g' pggb-x -i
sed 's,"$fmt" smoothxg,"$fmt" ~/tools/smoothxg/bin/smoothxg-e4c1149141ceb896740a8831a5e35fa65b7ba238,g' pggb-x -i
sed 's,"$fmt" odgi,"$fmt" ~/tools/odgi/bin/odgi-bfae0b3c70a0f044d7ce8510b7a45f59d5138626,g' pggb -i
mv pggb-x pggb-b3c55023313c28d89281d2e0ca0f10d25685d4e7
cd ..

wget -c https://github.com/vgteam/vg/releases/download/v1.40.0/vg
chmod +x vg

git clone --recursice https://github.com/pangenome/vcfbub.git
cd vcfbub
git checkout d92ae50da4926612f9bcf5d3ea8506a78a348734 # version 0.1.0
cargo build --release

/gnu/store/hkbkw85gjvsyqvx9vv9bw0ynmad989ag-vcflib-1.0.3+a36dbe9-11/bin/vcfwave


# MashMap3-based PGGB
mkdir forks && cd forks
git clone --recursive https://github.com/bkille/wfmash
cd wfmash
git checkout minmer-nominimizer && git pull && git submodule update --init --recursive
git checkout edbb5c3b678a0e379181f34295fd5a40bed95a17
cmake -H. -Bbuild && cmake --build build -- -j 48
mv build/bin/wfmash build/bin/wfmash-edbb5c3b678a0e379181f34295fd5a40bed95a17-mm3
cd ../..

cd pggb
git checkout master && git pull && git submodule update --init --recursive
git checkout b3c55023313c28d89281d2e0ca0f10d25685d4e7
cp pggb pggb-x
sed 's,"$fmt" wfmash,"$fmt" ~/tools/forks/wfmash/build/bin/wfmash-edbb5c3b678a0e379181f34295fd5a40bed95a17-mm3,g' pggb-x -i
sed 's,$(wfmash,$(~/tools/forks/wfmash/build/bin/wfmash-edbb5c3b678a0e379181f34295fd5a40bed95a17-mm3,g' pggb-x -i
sed 's,"$fmt" seqwish,"$fmt" ~/tools/seqwish/bin/seqwish-f362f6f5ea89dbb6a0072a8b8ba215e663301d33,g' pggb-x -i
sed 's,"$fmt" smoothxg,"$fmt" ~/tools/smoothxg/bin/smoothxg-e4c1149141ceb896740a8831a5e35fa65b7ba238,g' pggb-x -i
sed 's,"$fmt" odgi,"$fmt" ~/tools/odgi/bin/odgi-bfae0b3c70a0f044d7ce8510b7a45f59d5138626,g' pggb -i
mv pggb-x pggb-b3c55023313c28d89281d2e0ca0f10d25685d4e7-mm3
cd ..
```


## Data

```shell
mkdir -p $DIR_REPOSITORY/assemblies/sars_cov_2
cd $DIR_REPOSITORY/assemblies/sars_cov_2

cp $DIR_REPOSITORY/data/accession_numbers.SARS-CoV-2.txt .
cp $DIR_REPOSITORY/data/download_SARS-CoV-2.sh .
```
