# Preparation

Variables:

```shell
DIR_BASE=/lizardfs/guarracino/pggb-paper

WFMASH=/home/guarracino/tools/wfmash/build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0
```

## Data

### A. thaliana

66 assemblies from https://www.nature.com/articles/s41586-023-06062-z.

82 assemblies in total: on `2023/11/30`, specifying `Status: Latests` and `Exclude: Exclude partial` (`txid3702[Organism] AND (latest[filter] AND all[filter] NOT partial[filter])`), we downloaded 111 assemblies from GenBank (`https://www.ncbi.nlm.nih.gov/assembly`). We added GCA_028009825.1, as GCA_028009825.2 added two 45S rDNA for chr2 and chr4 in October, which is before we run DeepVariant. We kept those with at most 7 contigs (5 or 7), removing contigs representing mitochondria and plasmids in the only 3 assemblies that featured them.

### E. coli

4104 assemblies in total: on `2023/11/30`, specifying `Status: Latests`, `Assembly level: Complete genome` and `Exclude: Exclude partial` (`txid562[Organism] AND (latest[filter] AND "complete genome"[filter] AND all[filter] NOT partial[filter])`), we downloaded 4104 assemblies from GenBank (`https://www.ncbi.nlm.nih.gov/assembly`). From these, we randombly sampled 41 assemblies to built a pangenome 10X smaller.

```shell
samtools faidx ecoli4104.fa.gz $(grep -f <(cut -f 1 ecoli4104.fa.gz.fai -d '#' | sort | uniq | shuf | head -n 410) ecoli4104.fa.gz.fai | cut -f 1) | bgzip -@ 48 -l 9 > ecoli410.fa.gz && samtools faidx ecoli410.fa.gz
```

### H. sapiens

47 diploid assemblies (88 haplotypes) from the HPRC + GRCh38 and CHM13 reference genomes.

### Mouse

36 assemblies in total: on `2023/11/30`, specifying `Status: Latests`, `Assembly level: Chromosome` and `Exclude: Exclude anomalous` (`txid10090[Organism] AND (latest[filter] AND "chromosome level"[filter] AND all[filter] NOT anomalous[filter])`), we downloaded 36 assemblies from GenBank (`https://www.ncbi.nlm.nih.gov/assembly`).

```shell
ls *fna | while read FASTA; do ACC=$(basename $FASTA .fna | cut -f 1,2 -d '_'); echo $ACC; ./rename-mice-fasta.sh $FASTA $ACC | bgzip -l 9 -@ 48 > $ACC.fasta.gz; samtools faidx $ACC.fasta.gz; done

ls *fasta.gz | while read f; do echo $f; samtools faidx $f $(grep chr19 $f.fai | cut -f 1) >> mmusculus36.chr19.fa; done
bgzip -@ 48 -l 9 mmusculus36.chr19.fa && samtools faidx mmusculus36.chr19.fa.gz
```

### Primates

#### Release 2023/12/05

```shell
mkdir -p $DIR_BASE/assemblies/primates
cd $DIR_BASE/assemblies/primates
wget -c https://s3.amazonaws.com/genomeark/species/Gorilla_gorilla/mGorGor1/assembly_curated/mGorGor1.dip.cur.20231122.fasta.gz
wget -c https://s3.amazonaws.com/genomeark/species/Pan_paniscus/mPanPan1/assembly_curated/mPanPan1.dip.cur.20231122.fasta.gz
wget -c https://s3.amazonaws.com/genomeark/species/Pan_troglodytes/mPanTro3/assembly_curated/mPanTro3.dip.cur.20231122.fasta.gz
wget -c  https://s3.amazonaws.com/genomeark/species/Pongo_abelii/mPonAbe1/assembly_curated/mPonAbe1.dip.cur.20231205.fasta.gz
wget -c https://s3.amazonaws.com/genomeark/species/Pongo_pygmaeus/mPonPyg2/assembly_curated/mPonPyg2.dip.cur.20231122.fasta.gz
wget -c https://genomeark.s3.amazonaws.com/species/Symphalangus_syndactylus/mSymSyn1/assembly_curated/mSymSyn1.dip.cur.20231205.fasta.gz

# Apply PanSN-spec
for species in mGorGor1 mPanPan1; do
    zcat $species.dip.cur.20231122.fasta.gz | \
        sed "/^>/ s/chr\(.*\)_\(mat\|pat\)_*/$species##\2##\0/; s/#mat#/M/; s/#pat#/P/" | \
        bgzip -@ 48 -l 9 > $species.fa.gz && samtools faidx $species.fa.gz
done
for species in mPanTro3 mPonPyg2; do
    zcat $species.dip.cur.20231122.fasta.gz  | \
        sed "/^>/ s/chr\(.*\)_\(hap1\|hap2\)_*/$species##\2##\0/; s/#hap1#/1/; s/#hap2#/2/" | \
        bgzip -@ 48 -l 9 > $species.fa.gz && samtools faidx $species.fa.gz
done
for species in mPonAbe1 mSymSyn1; do
    zcat $species.dip.cur.20231205.fasta.gz  | \
        sed "/^>/ s/chr\(.*\)_\(hap1\|hap2\)_*/$species##\2##\0/; s/#hap1#/1/; s/#hap2#/2/" | \
        bgzip -@ 48 -l 9 > $species.fa.gz && samtools faidx $species.fa.gz
done
rm *fasta.gz

wget -c https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/HG002/assemblies/hg002v1.0.1.fasta.gz
zcat hg002v1.0.1.fasta | sed -e 's/^>chr\(.*\)_MATERNAL/>hg002#M#chr\1_MATERNAL/' \
    -e 's/^>chr\(.*\)_PATERNAL/>hg002#P#chr\1_PATERNAL/' \
    -e 's/^>chrEBV/>hg002#P#chrEBV/' \
    -e 's/^>chrM/>hg002#M#chrM/' | bgzip -@ 48 -l 9 > hg002v101.fa.gz && samtools faidx hg002v101.fa.gz
rm hg002v1.0.1.fasta.gz

zcat chm13v2.0.fa.gz grch38.fa.gz hg002v101.fa.gz mGorGor1.fa.gz mPanPan1.fa.gz mPanTro3.fa.gz mPonAbe1.fa.gz mPonPyg2.fa.gz mSymSyn1.fa.gz | bgzip -@ 48 -l 9 > primates16.20231205.fa.gz && samtools faidx primates16.20231205.fa.gz


# grch38#1#chr6	29722775	33089696  MHC
$WFMASH chm13v2.0.fa.gz mSymSyn1.fa.gz -m -t 48 -N > mSymSyn1.vs.chm13v2.mappings.N.paf
  grep 'chm13#1#chr6' mSymSyn1.vs.chm13v2.mappings.N.paf
  mSymSyn1#1#chr23_hap1   78312223        0       78312223        -       chm13#1#chr6    172126628       19925566        98237789        87      78312223        14      id:f:0.959062   kc:f:0.576732
  mSymSyn1#2#chr23_hap2   71145118        0       71145118        -       chm13#1#chr6    172126628       24314822        95459940        90      71145118        14      id:f:0.960383   kc:f:0.634831
  ...

cat \
  <(samtools faidx chm13v2.0.fa.gz chm13#1#chr6) \
  <(samtools faidx grch38.fa.gz grch38#1#chr6) \
  <(samtools faidx hg002v101.fa.gz hg002#M#chr6_MATERNAL hg002#P#chr6_PATERNAL) \
  <(samtools faidx mGorGor1.fa.gz mGorGor1#M#chr5_mat_hsa6 mGorGor1#P#chr5_pat_hsa6) \
  <(samtools faidx mPanPan1.fa.gz mPanPan1#M#chr5_mat_hsa6 mPanPan1#P#chr5_pat_hsa6) \
  <(samtools faidx mPanTro3.fa.gz mPanTro3#1#chr5_hap1_hsa6 mPanTro3#2#chr5_hap2_hsa6) \
  <(samtools faidx mPonAbe1.fa.gz mPonAbe1#1#chr5_hap1_hsa6 mPonAbe1#2#chr5_hap2_hsa6) \
  <(samtools faidx mPonPyg2.fa.gz mPonPyg2#1#chr5_hap1_hsa6 mPonPyg2#2#chr5_hap2_hsa6) \
  <(samtools faidx mSymSyn1.fa.gz mSymSyn1#1#chr23_hap1 mSymSyn1#2#chr23_hap2) | \
  bgzip -@ 48 -l 9 > primates16.hsa6.fa.gz && samtools faidx primates16.hsa6.fa.gz
```

### S. cerevisiae

ToDo

### Soy

ToDo

### Tomato

ToDo

## Tools

```shell
mkdir -p ~/tools $$ cd ~/tools

git clone --recursive https://github.com/waveygang/wfmash
cd wfmash
git checkout master && git pull && git submodule update --init --recursive
git checkout 0b191bb84ffdfd257354c1aa82a7f1e13dc536d0
cmake -H. -Bbuild && cmake --build build -- -j 48
cp build/bin/wfmash build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0
cd ..

git clone --recursive https://github.com/ekg/seqwish.git
cd seqwish
git checkout master && git pull && git submodule update --init --recursive
git checkout f44b402f0c2e02988d431d9b2e5eba9727cf93a9
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/seqwish bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9
cd ..

git clone --recursive https://github.com/pangenome/smoothxg.git
cd smoothxg
git checkout master && git pull && git submodule update --init --recursive
git checkout ae949a1053fb3d7c7fb7bdf358aefcbcbd8073a4
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/smoothxg bin/smoothxg-ae949a1053fb3d7c7fb7bdf358aefcbcbd8073a4
cd ..

git clone --recursive https://github.com/pangenome/odgi.git
cd odgi
git checkout master && git pull && git submodule update --init --recursive
git checkout 861b1c04f5622c5bb916a161c1abe812c213f1a5
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/odgi bin/odgi-861b1c04f5622c5bb916a161c1abe812c213f1a5
cd ..

clone https://github.com/marschall-lab/GFAffix.git
cd GFAffix
git checkout main && git pull && git submodule update --init --recursive
git checkout d630eb7d9827340f5f292e57cb3cb5e31e6f86f0
env -i bash -c 'PATH=:/usr/local/bin:/usr/bin:/bin ~/.cargo/bin/cargo build --release'
cp target/release/gfaffix target/release/gfaffix-d630eb7d9827340f5f292e57cb3cb5e31e6f86f0
cd ..

git clone --recursive https://github.com/pangenome/pggb.git
cd pggb
git checkout master && git pull && git submodule update --init --recursive
git checkout 13482bd06359a7ad8e3d3e0dd6eb6d9399f26046
cp pggb pggb-x
sed 's,"$fmt" wfmash,"$fmt" ~/tools/wfmash/build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0,g' pggb-x -i
sed 's,$(wfmash,$(~/tools/wfmash/build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0,g' pggb-x -i
sed 's,"$fmt" seqwish,"$fmt" ~/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9,g' pggb-x -i
sed 's,"$fmt" smoothxg,"$fmt" ~/tools/smoothxg/bin/smoothxg-ae949a1053fb3d7c7fb7bdf358aefcbcbd8073a4,g' pggb-x -i
sed 's,"$fmt" odgi,"$fmt" ~/tools/odgi/bin/odgi-861b1c04f5622c5bb916a161c1abe812c213f1a5,g' pggb-x -i
sed 's,"$fmt" gfaffix,"$fmt" ~/tools/GFAffix/target/release/gfaffix-d630eb7d9827340f5f292e57cb3cb5e31e6f86f0,g' pggb-x -i
mv pggb-x pggb-13482bd06359a7ad8e3d3e0dd6eb6d9399f26046
cp partition-before-pggb partition-before-pggb-x
sed 's,"$fmt" wfmash,"$fmt" ~/tools/wfmash/build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0,g' partition-before-pggb-x -i
sed 's,$(wfmash,$(~/tools/wfmash/build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0,g' partition-before-pggb-x -i
sed 's,"$fmt" seqwish,"$fmt" ~/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9,g' partition-before-pggb-x -i
sed 's,"$fmt" smoothxg,"$fmt" ~/tools/smoothxg/bin/smoothxg-ae949a1053fb3d7c7fb7bdf358aefcbcbd8073a4,g' partition-before-pggb-x -i
sed 's,"$fmt" odgi,"$fmt" ~/tools/odgi/bin/odgi-861b1c04f5622c5bb916a161c1abe812c213f1a5,g' partition-before-pggb-x -i
sed 's,"$fmt" gfaffix,"$fmt" ~/tools/GFAffix/target/release/gfaffix-d630eb7d9827340f5f292e57cb3cb5e31e6f86f0,g' partition-before-pggb-x -i
mv partition-before-pggb-x partition-before-pggb-13482bd06359a7ad8e3d3e0dd6eb6d9399f26046
cd ..

wget -c https://github.com/vgteam/vg/releases/download/v1.53.0/vg
chmod +x vg

git clone --recursice https://github.com/pangenome/vcfbub.git
cd vcfbub
git checkout 26a1f0cb216a423f8547c4ad0e0ce38cb9d324b9 # version 0.1.0
cargo build --release

/gnu/store/79r54wk4p3705dk89jg9hidyvf4754jp-vcflib-1.0.3+fdcdaad-10/bin/vcfwave

# Minigraph-Cactus
cd /lizardfs/guarracino/tools
wget -c https://github.com/ComparativeGenomicsToolkit/cactus/releases/download/v2.7.0/cactus-bin-v2.7.0.tar.gz
tar -xzf cactus-bin-v2.7.0.tar.gz && rm cactus-bin-v2.7.0.tar.gz
cd cactus-bin-v2.7.0

# guix install python-virtualenv
virtualenv -p python3 venv-cactus-v2.7.0
printf "export PATH=$(pwd)/bin:\$PATH\nexport PYTHONPATH=$(pwd)/lib:\$PYTHONPATH\n" >> venv-cactus-v2.7.0/bin/activate
source venv-cactus-v2.7.0/bin/activate
python3 -m pip install -U setuptools pip wheel
python3 -m pip install -U .
python3 -m pip install -U -r ./toil-requirement.txt

cactus ./jobstore ./examples/evolverMammals.txt ./evolverMammals.hal --realTimeLogging # test
```
