# Preparation

Variables:

```shell
DIR_BASE=/lizardfs/guarracino/pggb-paper

WFMASH=/home/guarracino/tools/wfmash/build/bin/wfmash-2b42cba6a6623d7838605679dbce73b6fc2bde33
```

## Data

### A. thaliana

81 assemblies in total: on `2023/11/30`, specifying `Status: Latests` and `Exclude: Exclude partial` (`txid3702[Organism] AND (latest[filter] AND all[filter] NOT partial[filter])`), we downloaded 111 assemblies from GenBank (`https://www.ncbi.nlm.nih.gov/assembly`). We kept those with at most 7 contigs (5 or 7), removing contigs representing mitochondria and plasmids in the only 3 assemblies that featured them.

### E. coli

4104 assemblies in total: on `2023/11/30`, specifying `Status: Latests`, `Assembly level: Complete genome` and `Exclude: Exclude partial` (`txid562[Organism] AND (latest[filter] AND "complete genome"[filter] AND all[filter] NOT partial[filter])`), we downloaded 4104 assemblies from GenBank (`https://www.ncbi.nlm.nih.gov/assembly`).

### H. sapiens
hsapiens90.chr6 15508376475 81.17 1183.33 135.52 143972 0.971475
### Mouse
mouse17.chr19 994731502 11.52 203.80 29.48 223951 0.907288

### Primates

#### Release 2023/11/22

```shell
mkdir -p $DIR_BASE/assemblies/primates
cd $DIR_BASE/assemblies/primates
wget -c https://s3.amazonaws.com/genomeark/species/Gorilla_gorilla/mGorGor1/assembly_curated/mGorGor1.dip.cur.20231122.fasta.gz
wget -c https://s3.amazonaws.com/genomeark/species/Pan_paniscus/mPanPan1/assembly_curated/mPanPan1.dip.cur.20231122.fasta.gz
wget -c https://s3.amazonaws.com/genomeark/species/Pan_troglodytes/mPanTro3/assembly_curated/mPanTro3.dip.cur.20231122.fasta.gz
wget -c https://s3.amazonaws.com/genomeark/species/Pongo_abelii/mPonAbe1/assembly_curated/mPonAbe1.dip.cur.20231122.fasta.gz
wget -c https://s3.amazonaws.com/genomeark/species/Pongo_pygmaeus/mPonPyg2/assembly_curated/mPonPyg2.dip.cur.20231122.fasta.gz
wget -c https://s3.amazonaws.com/genomeark/species/Symphalangus_syndactylus/mSymSyn1/assembly_curated/mSymSyn1.dip.cur.20231122.fasta.gz

# Apply PanSN-spec
for species in mGorGor1 mPanPan1; do
    zcat $species.dip.cur.20231122.fasta.gz | \
        sed "/^>/ s/chr\(.*\)_\(mat\|pat\)_*/$species##\2##\0/; s/#mat#/M/; s/#pat#/P/" | \
        bgzip -@ 48 -l 9 > $species.fa.gz && samtools faidx $species.fa.gz
done
for species in mPanTro3 mPonAbe1 mSymSyn1 mPonPyg2; do
    zcat $species.dip.cur.20231122.fasta.gz  | \
        sed "/^>/ s/chr\(.*\)_\(hap1\|hap2\)_*/$species##\2##\0/; s/#hap1#/1/; s/#hap2#/2/" | \
        bgzip -@ 48 -l 9 > $species.fa.gz && samtools faidx $species.fa.gz
done
rm *fasta.gz

wget -c https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/HG002/assemblies/hg002v1.0.1.fasta.gz
zcat hg002v1.0.1.fasta | sed -e 's/^>chr\(.*\)_MATERNAL/>hg002#M#chr\1_MATERNAL/' \
    -e 's/^>chr\(.*\)_PATERNAL/>hg002#P#chr\1_PATERNAL/' \
    -e 's/^>chrEBV/>hg002#P#chrEBV/' \
    -e 's/^>chrM/>hg002#M#chrM/' | bgzip -@ 48 -l 9 > hg002v101.fa.gz && samtools faidx hg002v101.fa.gz
rm hg002v1.0.1.fasta.gz

zcat chm13v2.0.fa.gz grch38.fa.gz hg002v101.fa.gz mGorGor1.fa.gz mPanPan1.fa.gz mPanTro3.fa.gz mPonAbe1.fa.gz mPonPyg2.fa.gz mSymSyn1.fa.gz | bgzip -@ 48 -l 9 > primates16.20231129.fa.gz && samtools faidx primates16.20231129.fa.gz


# grch38#1#chr6	29722775	33089696  MHC
$WFMASH chm13v2.0.fa.gz mSymSyn1.fa.gz -m -t 48 -N > mSymSyn1.vs.chm13v2.mappings.N.paf
$WFMASH chm13v2.0.fa.gz mSymSyn1.fa.gz -m -t 48 > mSymSyn1.vs.chm13v2.mappings.paf
grep 'chm13#1#chr6' mSymSyn1.vs.chm13v2.mappings.N.paf
  mSymSyn1#1#chr23_hap1   78312223        0       78312223        -       chm13#1#chr6    172126628       19925566        98237789        87      78312223        14      id:f:95.9062    kc:f:0.576732
  mSymSyn1#2#chr23_hap2   71145118        0       71145118        -       chm13#1#chr6    172126628       24314822        95459940        90      71145118        14      id:f:96.0383    kc:f:0.634831
  ...

cat \
  <(samtools faidx chm13v2.0.fa.gz chm13#1#chr6) \
  <(samtools faidx grch38.fa.gz grch38#1#chr6) \
  <(samtools faidx hg002v101.fa.gz hg002#M#chr6_MATERNAL hg002#P#chr6_PATERNAL) \
  <(samtools faidx mGorGor1.fa.gz mGorGor1#M#chr5_mat_hsa6 mGorGor1#P#chr5_pat_hsa6) \
  <(samtools faidx mPanPan1.fa.gz mPanPan1#M#chr5_mat_hsa6 mPanPan1#P#chr5_pat_hsa6) \
  <(samtools faidx mPanTro3.fa.gz mPanTro3#1#chr5_hap1_hsa6 mPanTro3#2#chr5_hap2_hsa6) \
  <(samtools faidx mPonAbe1.fa.gz mPonAbe1#1#chr5_hap1_hsa6 mPonAbe1#2#chr5_hap2_hsa6) \
  <(samtools faidx mPonPyg2.fa.gz mPonPyg2#1#chr5_hap1_hsa6 mPonPyg2#2#chr5_hap2_hsa6) \
  <(samtools faidx mSymSyn1.fa.gz mSymSyn1#1#chr23_hap1 mSymSyn1#2#chr23_hap2) | \
  bgzip -@ 48 -l 9 > primates16.hsa6.fa.gz && samtools faidx primates16.hsa6.fa.gz -f
```

#### Release 2022/11/11

Download:

```shell
mkdir -p $DIR_BASE/assemblies

# T2T.primates_assemblies.urls.txt refers to assemblies in https://genomeark.github.io/t2t-draft-assembly/, but they were taken from https://genomeark.s3.amazonaws.com/index.html?prefix=species/ (11 January 2023)
# https://docs.google.com/spreadsheets/d/1bWZ1SjCn6I34QMqXeg-zwzK7D-EDizw1GQCkQ_Ily6E/edit#gid=0
sbatch -c 1 --wrap "cd $DIR_BASE/assemblies; (echo https://hgdownload.soe.ucsc.edu/goldenPath/hs1/bigZips/hs1.fa.gz; cat ../data/T2T.primates_assemblies.urls.txt) | parallel -j 8 'wget -q {} && echo got {}'"
mv hs1.fa.gz chm13v2.fasta.gz

# Apply PanSN-spec
ls *fasta.gz | while read f; do
  echo $f
  prefix=$(echo $f | cut -f 1 -d '.')

  $RUN_FASTIX -p "${prefix}#" <(zcat $f ) | sed -e 's/haplotype//g' -e 's/-/#/g' -e 's/unassigned/U/g' | bgzip -c -@ 48 > $prefix.fa.gz
  samtools faidx $prefix.fa.gz
done
```

Map contigs against the references:

```shell
mkdir -p $DIR_BASE/partitioning
cd $DIR_BASE/partitioning

REFS=/lizardfs/guarracino/chromosome_communities/assemblies/chm13v2+grch38masked.fa.gz

ls $DIR_BASE/assemblies/*.fa.gz | while read FASTA; do
  SPECIES=$(basename $FASTA .fa.gz | cut -f 1,2 -d '.');
  echo $SPECIES
  
  PAF=$DIR_BASE/partitioning/$SPECIES.vs.ref.p90.paf
  sbatch -p workers -c 20 --wrap "$RUN_WFMASH -t 20 -m -N -s 50k -l 150k -p 90 -H 0.001 $REFS $FASTA > $PAF"
done
```

Collect unmapped contigs and remap them in split mode:

```shell
REFS=/lizardfs/guarracino/chromosome_communities/assemblies/chm13v2+grch38masked.fa.gz

(ls $DIR_BASE/assemblies/*.fa.gz | grep 'chm13\|primates' -v) | while read FASTA; do
  SPECIES=$(basename $FASTA .fa.gz | cut -f 1,2 -d '.');
  echo $SPECIES
  
  UNALIGNED=$DIR_BASE/partitioning/$SPECIES.unaligned
  
  PAF=$DIR_BASE/partitioning/$SPECIES.vs.ref.p90.paf
  comm -23 <(cut -f 1 $FASTA.fai | sort) <(cut -f 1 $PAF | sort) > $UNALIGNED.txt
  if [[ $(wc -l $UNALIGNED.txt | cut -f 1 -d ' ' ) != 0 ]];
  then 
    samtools faidx $FASTA $(tr '\n' ' ' < $UNALIGNED.txt) > $UNALIGNED.fa
    samtools faidx $UNALIGNED.fa
    sbatch -p workers -c 20 --wrap "$RUN_WFMASH -t 20 -m -s 50k -l 150k -p 90 -H 0.001 $REFS $UNALIGNED.fa > $UNALIGNED.split.vs.ref.p90.paf"
  fi
done
```

Collect our best mapping for each of our attempted split rescues:

```shell
ls *.unaligned.split.vs.ref.p90.paf | while read PAF; do
  cat $PAF | awk -v OFS='\t' '{ print $1,$11,$0 }' | sort -n -r -k 1,2 | \
    awk -v OFS='\t' '$1 != last { print($3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15); last = $1; }'
done > rescues.paf
```

Collect partitioned contigs:

```shell
# Take only haplotype-assigned contigs
( seq 6 7 8 ) | while read i; do cat *paf | grep -P -e "[chm13|grch38]#chr$i\t" | grep '#U#' -v | cut -f 1 | sort | uniq | awk '{print($0"$")}' > chr$i.contigs.txt; done

( seq 6 7 8  ) | while read i; do
    echo chr$i

    rm chr$i.fa*
    (ls $DIR_BASE/assemblies/*.fa.gz | grep 'chm13\|primates' -v) | while read FASTA; do
      echo $FASTA

      samtools faidx $FASTA $(cut -f 1 $FASTA.fai | grep -f chr$i.contigs.txt) >> chr$i.fa
    done
done

REFS=/lizardfs/guarracino/chromosome_communities/assemblies/chm13v2+grch38masked.fa.gz

# Only chr 6 and 7 are available from primates4
( seq 6 7 ) | while read i; do
    echo chr$i
    # 4 haplotypes (from primates4) + 12 haplotypes (6 diploid assemblies) + 1 haplotype (chm13#chr$i)
    cat \
      chr$i.fa \
      <(zcat /lizardfs/guarracino/pggb-paper/assemblies/primates4.chr$i.fa.gz) \
      <(samtools faidx $REFS $(grep chm13#chr$i $REFS.fai | cut -f 1)) \
      > primates17.chr$i.fa
    bgzip -@ 48 primates17.chr$i.fa
    samtools faidx primates17.chr$i.fa.gz
done

( seq 8 8 ) | while read i; do
    echo chr$i
    # 12 haplotypes (6 diploid assemblies) + 2 haplotype (chm13#chr$i and grch38#chr$i)
    cat \
      chr$i.fa \
      <(samtools faidx $REFS $(grep hr$i $REFS.fai | cut -f 1)) \
      > primates14.chr$i.fa
    bgzip -@ 48 primates14.chr$i.fa
    samtools faidx primates14.chr$i.fa.gz
done
```

### S. cerevisiae


scerevisiae8 96255507 6.47 8.78 3.53 53742 0.968729
scerevisiae142 1702093905 55.29 1021.81 112.98 62796 0.955988
scerevisiae142* 1702093905 41.41 562.89 75.91 62580 0.955650


### Soy
soy37.chr18 2240871558 20.50 599.66 29.17 101907 0.907878

### Tomato
tomato23.chr2 1280460312 20.69 78.53 43.84 39243 0.948173


## Tools

```shell
mkdir -p ~/tools $$ cd ~/tools

git clone --recursive https://github.com/waveygang/wfmash
cd wfmash
git checkout master && git pull && git submodule update --init --recursive
git checkout ecdfb95e660eda46c3c51312a7db54394270271a
cmake -H. -Bbuild && cmake --build build -- -j 48
cp build/bin/wfmash build/bin/wfmash-ecdfb95e660eda46c3c51312a7db54394270271a
cd ..

git clone --recursive https://github.com/ekg/seqwish.git
cd seqwish
git checkout master && git pull && git submodule update --init --recursive
git checkout f44b402f0c2e02988d431d9b2e5eba9727cf93a9
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/seqwish bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9
cd ..

git clone --recursive https://github.com/pangenome/smoothxg.git
cd smoothxg
git checkout master && git pull && git submodule update --init --recursive
git checkout aaa0b283e13ca57c4e6e4e67a03451925f5342f1
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/smoothxg bin/smoothxg-aaa0b283e13ca57c4e6e4e67a03451925f5342f1
cd ..

git clone --recursive https://github.com/pangenome/odgi.git
cd odgi
git checkout master && git pull && git submodule update --init --recursive
git checkout fbdb4d23a512cdcbede21c7a453f62415a2ca7e6
rm -rf build/
cmake -H. -Bbuild && cmake --build build -- -j 48
cp bin/odgi bin/odgi-fbdb4d23a512cdcbede21c7a453f62415a2ca7e6
cd ..

clone https://github.com/marschall-lab/GFAffix.git
cd GFAffix
git checkout main && git pull && git submodule update --init --recursive
git checkout d630eb7d9827340f5f292e57cb3cb5e31e6f86f0
env -i bash -c 'PATH=:/usr/local/bin:/usr/bin:/bin ~/.cargo/bin/cargo build --release'
cp target/release/gfaffix target/release/gfaffix-d630eb7d9827340f5f292e57cb3cb5e31e6f86f0
cd ..

git clone --recursive https://github.com/pangenome/pggb.git
cd pggb
git checkout master && git pull && git submodule update --init --recursive
git checkout b26786adce800bbf2664e41e4c95d57cb3ee048e
cp pggb pggb-x
sed 's,"$fmt" wfmash,"$fmt" ~/tools/wfmash/build/bin/wfmash-ecdfb95e660eda46c3c51312a7db54394270271a,g' pggb-x -i
sed 's,$(wfmash,$(~/tools/wfmash/build/bin/wfmash-ecdfb95e660eda46c3c51312a7db54394270271a,g' pggb-x -i
sed 's,"$fmt" seqwish,"$fmt" ~/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9,g' pggb-x -i
sed 's,"$fmt" smoothxg,"$fmt" ~/tools/smoothxg/bin/smoothxg-aaa0b283e13ca57c4e6e4e67a03451925f5342f1,g' pggb-x -i
sed 's,"$fmt" odgi,"$fmt" ~/tools/odgi/bin/odgi-fbdb4d23a512cdcbede21c7a453f62415a2ca7e6,g' pggb-x -i
sed 's,"$fmt" gfaffix,"$fmt" ~/tools/GFAffix/target/release/gfaffix-d630eb7d9827340f5f292e57cb3cb5e31e6f86f0,g' pggb-x -i
mv pggb-x pggb-b26786adce800bbf2664e41e4c95d57cb3ee048e
cp partition-before-pggb partition-before-pggb-x
sed 's,"$fmt" wfmash,"$fmt" ~/tools/wfmash/build/bin/wfmash-ecdfb95e660eda46c3c51312a7db54394270271a,g' partition-before-pggb-x -i
sed 's,$(wfmash,$(~/tools/wfmash/build/bin/wfmash-ecdfb95e660eda46c3c51312a7db54394270271a,g' partition-before-pggb-x -i
sed 's,"$fmt" seqwish,"$fmt" ~/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9,g' partition-before-pggb-x -i
sed 's,"$fmt" smoothxg,"$fmt" ~/tools/smoothxg/bin/smoothxg-aaa0b283e13ca57c4e6e4e67a03451925f5342f1,g' partition-before-pggb-x -i
sed 's,"$fmt" odgi,"$fmt" ~/tools/odgi/bin/odgi-fbdb4d23a512cdcbede21c7a453f62415a2ca7e6,g' partition-before-pggb-x -i
sed 's,"$fmt" gfaffix,"$fmt" ~/tools/GFAffix/target/release/gfaffix-d630eb7d9827340f5f292e57cb3cb5e31e6f86f0,g' partition-before-pggb-x -i
mv partition-before-pggb-x partition-before-pggb-b26786adce800bbf2664e41e4c95d57cb3ee048e
cd ..

wget -c https://github.com/vgteam/vg/releases/download/v1.52.0/vg
chmod +x vg

git clone --recursice https://github.com/pangenome/vcfbub.git
cd vcfbub
git checkout d92ae50da4926612f9bcf5d3ea8506a78a348734 # version 0.1.0
cargo build --release

/gnu/store/hkbkw85gjvsyqvx9vv9bw0ynmad989ag-vcflib-1.0.3+a36dbe9-11/bin/vcfwave


# Minigraph-Cactus
cd /lizardfs/guarracino/tools
wget -c https://github.com/ComparativeGenomicsToolkit/cactus/releases/download/v2.6.13/cactus-bin-v2.6.13.tar.gz
tar -xzf cactus-bin-v2.6.13.tar.gz
cd cactus-bin-v2.6.13

# guix install python-virtualenv
virtualenv -p python3 venv-cactus-v2.6.13
printf "export PATH=$(pwd)/bin:\$PATH\nexport PYTHONPATH=$(pwd)/lib:\$PYTHONPATH\n" >> venv-cactus-v2.6.13/bin/activate
source venv-cactus-v2.6.13/bin/activate
python3 -m pip install -U setuptools pip wheel
python3 -m pip install -U .
python3 -m pip install -U -r ./toil-requirement.txt
```
