# Pangenome building

## All species

Variables:

```shell
RUN_PGGB=/home/guarracino/tools/pggb/pggb-b3c55023313c28d89281d2e0ca0f10d25685d4e7
RUN_ODGI=/home/guarracino/tools/odgi/bin/odgi-19cea997e29ea590c1ccd75c1d039b2b1ed44f74
DIR_GRAPHS=/lizardfs/guarracino/pggb-paper/graphs
DIR_EVALUATIONS=/lizardfs/guarracino/pggb-paper/evaluations
PATH_PANGENOMES_TSV=/lizardfs/guarracino/pggb-paper/data/pangenomes.tsv
RUN_GFA_2_EVALUATION=/lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh
THREADS=48
```

Run `pggb`:

```shell
mkdir -p $DIR_GRAPHS
cd $DIR_GRAPHS

sed '1d' $PATH_PANGENOMES_TSV | grep rat -v | sort -k 4n | while read f p s n k G POA O REF; do
  seq 3 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;
    
    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    echo sbatch -c $THREADS -p 386mem -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $RUN_PGGB -i $f -p $p -s $s -n $n -k $k -P $POA -O $O -G $G -t $THREADS -o $out; mv $out $DIR_GRAPHS/$out;"
  done
done

# Script to parse statistics
(echo dataset tool runtime_s memory_kb replicate | tr ' ' '\t'; grep 'wfmash-\|seqwish-\|smoothxg-\|odgi-\|gfaffix[[:space:]]' */*log -A 1  | grep 'wfmash::align' -v | grep '^--$' -v | python3 ../scripts/pggb_stats.py | awk -v OFS='\t' '{print($0, substr($1, length($1), 1))}') > stats.tsv

#Old
#grep 'wfmash-\|seqwish-\|smoothxg-\|odgi-' hsapiens90.chr6.masked_p98.s10000.n90.k79.G700-900-1100.Pasm5.O0001/*log -A 1  | grep 'wfmash::align' -v | grep '^--$' -v | cut -f 1,2 -d ' ' | rev | cut -f 1 -d '/' | rev | sed 's/-[^ ]*//g' | sed 's/s user//g' | paste - - | sed 's/ \([[:alpha:]]\)/_\1/g' | column -t

# Risky verbose way
#ls */*log | while read f; do x=$(echo $f | cut -f 1 -d '.'); grep 'max memory' $f | head -n 4 | cut -f 7 -d ' ' | sed 's/s//g' | awk '{print($0/60/60)}' | paste - - - - | awk '{print($1,"wfmash-mapping", $2, #"wfmash-alignment",$3,"seqwish",$4,"smoothxg")}' | tr ' ' '\n' | paste - - | awk -v set=$x '{print($0,set)}' | column -t; done | column -t
```

Visualization by merging contigs by haplotype:

```shell
cd $DIR_GRAPHS

ls */*.final.og | grep '_1' | while read PATH_OG; do
  DATASET=$(basename "$PATH_OG" .smooth.final.og | cut -f 1 -d '.')
  echo $DATASET
  if [[ ! -s $PATH_OG.prefixes.txt ]]; then
    $RUN_ODGI paths -i $PATH_OG -L | cut -f 1,2 -d '#' | sort | uniq > $PATH_OG.prefixes.txt 
  fi

  #if [[ ! -s $PATH_OG.merged_by_haplotype.png ]]; then
    sbatch -c $THREADS -p workers -J $DATASET --wrap "hostname; $RUN_ODGI viz -i $PATH_OG -o $PATH_OG.merged_by_haplotype.z.png -x 1000 -y 500 -a 10 -z -I Consensus_ -M $PATH_OG.prefixes.txt"
    sbatch -c $THREADS -p workers -J $DATASET --wrap "hostname; $RUN_ODGI viz -i $PATH_OG -o $PATH_OG.merged_by_haplotype.du.png -x 1000 -y 500 -a 10 -du -I Consensus_ -M $PATH_OG.prefixes.txt"
  #fi
done
```

Evaluation:

```shell
mkdir -p $DIR_EVALUATIONS
cd $DIR_EVALUATIONS

sed '1d' $PATH_PANGENOMES_TSV | grep primates | grep rat -v | sort -k 4n | while read f p s n k G POA O REF; do
  seq 1 1 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;

    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    PATH_GFA=$DIR_GRAPHS/$out/*.final.gfa
    PATH_GFA=$(eval echo $PATH_GFA) # force '*' expansion
    echo sbatch -c $THREADS -p workers -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $RUN_GFA_2_EVALUATION $PATH_GFA $REF $DIR_EVALUATIONS/$out $THREADS;"
  done
done

# All results
((echo -n replicate"\t"dataset"\t"; cat */statistics*.tsv | head -n 1); ls */statistics.haplo.tsv | while read f; do name=$(echo $f | cut -f 1 -d '/'); sed '1d' $f | awk -v name=$name -v OFS='\t' '{ print(name,$0) }' | awk -v OFS='\t' '{print(substr($1, length($1), 1),$0)}'; done) > nucmer-based_evaluation.tsv

# Average results
(cat */statistics*.tsv | head -n 1; ls */statistics.haplo.tsv | while read f; do name=$(echo $f | cut -f 1 -d '/'); sed '1d' $f | awk -v name=$name '{ for (i=2; i<=NF; i++) {sum[i] += $i} } END { printf(name); for (i=2; i<=NF; i++) printf("\t%.6f ", sum[i]/NR); printf("\n") }'; done) | sed 's/haplotype/dataset/g' > nucmer-based_evaluation.mean.tsv
```

### Primates

Build graphs:

```shell
cd $DIR_GRAPHS

echo 14 | tr ' ' '\n' | while read n; do
  seq 6 6 | while read i; do
    echo $n $i
    f=/lizardfs/guarracino/tree_of_life_alignment/partitioning/primates$n.chr$i.fa.gz
    out=primates$n.chr${i}_p90.s5000.n$n.k79.G700-900-1100.Pasm5.O0001_1

    sbatch -c 48 -p 386mem -J primates$n.chr$i --wrap "hostname; cd /scratch; /home/guarracino/tools/pggb/pggb-b3c55023313c28d89281d2e0ca0f10d25685d4e7 -i $f -p 90 -s 5000 -n $n -k 79 -P asm5 -O 0.001 -G 700,900,1100 -t $THREADS -o $out; mv $out /lizardfs/guarracino/pggb-paper/graphs/$out;"
  done
done

Evaluation:

```shell
/lizardfs/guarracino/pggb-paper/assemblies/primates4.chr6.fa.gz 95 5000 4 79 700,900,1100 asm5 0.001 GRC38 1
sbatch -c 48 -p workers -J primates14.chr6 --wrap "hostname; cd /scratch; /lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh /lizardfs/guarracino/pggb-paper/graphs/primates14.chr6_p90.s5000.n14.k79.G700-900-1100.Pasm5.O0001_1/primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.gfa GRC38 /lizardfs/guarracino/pggb-paper/evaluations/primates14.chr6_p90.s5000.n14.k79.G700-900-1100.Pasm5.O0001_1 48;"
```

Add annotation:

```shell
cd $DIR_GRAPHS/primates14.chr6_p90.s5000.n12.k79.G700-900-1100.Pasm5.O0001_1/

# TO FIX: chm13#chr6 does not follow PanSN and it is annoying
sed 's/#chr/#1#chr/g' primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.gfa > primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.gfa
$RUN_ODGI build -g primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.gfa -o primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.og -t $THREADS -P

$RUN_ODGI inject -i primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.og -b /lizardfs/guarracino/pggb-paper/data/chr6.annotations.bed -o primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.inject.og -t $THREADS -P
```

## Notes

To remember that we are still doing this (both in PGGB and in GFA2EVALUATION) because of bugs in `vcfwave`:

```shell
# The TYPE info sometimes is wrong/missing
# There are variants without the ALT allele
bcftools annotate -x INFO/TYPE "$PATH_WAVED_VCF".gz  | awk '$5 != "."' | bgzip -@ THREADS -c > "$PATH_WAVED_FIXED_VCF".gz
```
