# Pangenome building

Variables:

```shell
DIR_BASE=/lizardfs/guarracino/pggb-paper
DIR_GRAPHS=/lizardfs/guarracino/pggb-paper/graphs
DIR_GRAPHS_MC=/lizardfs/guarracino/pggb-paper/graphs_mc
DIR_EVALUATIONS=/lizardfs/guarracino/pggb-paper/evaluations
PATH_PANGENOMES_TSV=/lizardfs/guarracino/pggb-paper/data/pangenomes.tsv

PGGB=/home/guarracino/tools/pggb/pggb-13482bd06359a7ad8e3d3e0dd6eb6d9399f26046
SEQWISH=/home/guarracino/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9
ODGI=/home/guarracino/tools/odgi/bin/odgi-861b1c04f5622c5bb916a161c1abe812c213f1a5
WFMASH=/home/guarracino/tools/wfmash/build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0

THREADS=48
```

## PGGB

Run `pggb`:

```shell
mkdir -p $DIR_GRAPHS
cd $DIR_GRAPHS

sed '1d' $PATH_PANGENOMES_TSV | sort -k 4n | head -n 1 | while read f p s n k REF; do
  seq 1 1 | while read c; do
    echo $f $p $s $n $k $REF $c;
    
    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k${k}_$c
    sbatch -c $THREADS -p allnodes -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $PGGB -i $f -p $p -s $s -n $n -k $k -P $POA -O $O -G $G -t $THREADS -o $out; mv $out $DIR_GRAPHS/$out;"
  done
done

#athaliana82  p95.s5000.n82.k47
# - minutes and kbs: 12244.090000  130590048.000000
# - graph.len, pangenome.len, ratio: 304629160  11052471644  36.2817
# - nucmer-based evaluation: 485379.925000  485379.925000  35389.425000  93172.837500  0.933970  0.845988  0.887598  616766.300000  612962.812500  0.363970  0.818340  520769.350000
# - DV-based evaluation: 
#   - all     158306.8646     153549.5046     9534.9385       13332.8646      0.9416  0.9236  0.9324
#   - easy    129634.5292     125721.9292     7408.2492       10719.4523      0.9454  0.9249  0.9350
#   - hard    28666.1662      27823.3477      2158.6523       2637.2338       0.9270  0.9158  0.9210

#hsapiens90.chr6.masked_p98.s5000.n90.k23
# - minutes and kbs: 1108.13 137716740
# - graph.len, pangenome.len, ratio: 178502732 15508376475      86.8803
# - nucmer-based evaluation: 139991.494382  139991.494382  4808.775281  3410.696629  0.966794  0.976253  0.971494  147086.213483  176923.494382  0.966858  0.818263  144800.269663
# - DV-based evaluation:
#   - all     315263.9545     314752.9091     17371.0227      16795.9318      0.9477  0.9496  0.9486
#   - easy    206729.1818     205451.6136     1515.0000       1354.1364       0.9928  0.9936  0.9932
#   - hard    108546.7955     109300.6136     16023.8182      15573.7727      0.8721  0.8748  0.8734

#tomato23  p95.s5000.n23.k17
# - minutes and kbs: 1340.710700  42056680.000000
# - graph.len, pangenome.len, ratio: 947468149  18691283883  19.7276
# - nucmer-based evaluation: 770435.863636   770435.863636   41592.409092    56267.909092    0.959553        0.940837        0.950046        882757.545456   1506867.363639  0.957112        0.619686        812028.272727
# - DV-based evaluation:
#   - all     209637.3889     204972.1111     19301.4444      3600.1250       0.9188  0.9801  0.9478
#   - easy    50210.6111      49080.1806      5784.8472       961.5417        0.9060  0.9808  0.9417
#   - hard    159454.9306     155925.5000     13771.4028      2672.1667       0.9207  0.9785  0.9478

#ecoli50_p90.s5000.n50.k17
# - minutes and kbs: 55.9888  19931880
# - graph.len, pangenome.len, ratio: 22226671   255615162   11.5004
# - nucmer-based evaluation: 51832.142857  51832.142857  438.755102  5374.591837  0.991904  0.907102  0.947598  58626.591837  53264.571429  0.971406  0.980149  52270.897959

#ecoli500_p90.s5000.n500.k17.xauto
# - minutes and kbs: 2483.92  210870380
# - graph.len, pangenome.len, ratio: 107921401  2562798947  23.7469
# - nucmer-based evaluation: 55747.088176  55747.088176  365.000000  7173.040080  0.993523  0.886785  0.936919  64540.587174  57134.865731  0.948857  0.980074  56112.088176

#primates16.hsa6_p95.s5000.n16.k47
# - minutes and kbs: 336.238 37047472
# - graph.len, pangenome.len, ratio: 327830132 2661626374      8.11892
# - nucmer-based evaluation: 1906489.000000  1906489.000000  184390.666667   230060.333333   0.921873        0.913233        0.917340        2143792.466667  2288443.666667  0.957722        0.891728        2090879.666667

#scerevisiae142_p95.s5000.n142.k47_1        1227.79  119680768
#scerevisiae142_p95.s5000.n142.k47_1.xauto  o  606.825  80253880
#scerevisiae142_p95.s5000.n142.k47_1        29801189  1702093905  57.115
#scerevisiae142_p95.s5000.n142.k47_1.xauto  37679369  1702093905  45.1731
#scerevisiae142_p95.s5000.n142.k47_1        58105.666667  58105.666667  873.070922  4593.028369  0.985282  0.929422  0.956373  64925.553191  61481.602837  0.944116  0.949184  58978.737589
#scerevisiae142_p95.s5000.n142.k47_1.xauto  57954.134752  57954.134752  852.503546  4537.503546  0.985840  0.929747  0.956822  64925.553191  61411.475177  0.941304  0.946332  58806.638298

#mmusculus36.chr19_p98.s5000.n36.k47_1
# - minutes and kbs: 228.079  28656920
# - graph.len, pangenome.len, ratio: 184097181  2150650973  11.6822
# - nucmer-based evaluation: 142647.200000  142647.200000  11617.514286  16777.085714  0.932760  0.900417  0.916186  161065.628571  169449.714286  0.916467  0.825424  154264.714286

#soy37  p95.s10000.n37.k23
# - minutes and kbs: 5984.393000   37538412.000000
# - graph.len, pangenome.len, ratio: 1705315400  36673233537  21.5052
# - nucmer-based evaluation: 1129570.111112  1129570.111112  88074.444445    107061.777778   0.928527        0.911535        0.919866        1264220.083333  2007884.472223  0.960226        0.634714        1217644.555554 


cd $DIR_GRAPHS
# Script to parse statistics
(echo dataset tool runtime_s memory_kb replicate | tr ' ' '\t'; grep 'wfmash-\|seqwish-\|smoothxg-\|odgi-\|gfaffix-' */*log -A 1  | grep 'wfmash::align' -v | grep '^--$' -v | python3 ../scripts/pggb_stats.py | awk -v OFS='\t' '{print($0, substr($1, length($1), 1))}') > runtime+memory.stats.tsv

awk 'BEGIN { FS="\t"; OFS="\t" } 
  NR>1 { runtime[$1,$5]+=$3; memory[$1,$5]=(memory[$1,$5]>$4)?memory[$1,$5]:$4 } 
  END { 
      print "dataset", "replicate", "total_runtime_m", "max_memory_kb"; 
      for (key in runtime) { 
          split(key, indices, SUBSEP); 
          print indices[1], indices[2], runtime[key]/60, memory[key] 
      } 
  }' runtime+memory.stats.tsv > runtime+memory.stats.grouped.tsv

cat runtime+memory.stats.tsv | grep -v 'odgi-viz\|odgi-layout\|odgi-draw' | awk 'BEGIN { FS="\t"; OFS="\t" } 
  NR>1 { runtime[$1,$5]+=$3; memory[$1,$5]=(memory[$1,$5]>$4)?memory[$1,$5]:$4 } 
  END { 
      print "dataset", "replicate", "total_runtime_m", "max_memory_kb"; 
      for (key in runtime) { 
          split(key, indices, SUBSEP); 
          print indices[1], indices[2], runtime[key]/60, memory[key] 
      } 
  }' > runtime+memory.stats.grouped.no-viz.tsv


grep 'athaliana\|tomato\|soy' runtime+memory.stats.grouped.no-viz.tsv | awk '
    BEGIN { FS=OFS="\t" }
    {
        split($1, arr1, ".")
        split($1, arr2, "_")
        key = arr1[1] FS arr2[2];  # Create a key for grouping
        sum[key] += $3;  # Sum up the 3rd column

        # Track the maximum of the 4th column
        if (!max[key] || $4 > max[key]) {
            max[key] = $4;
        }
    }
    END {
        # Print header
        print "pangenome", "parameters", "sum_3rd_column", "max_4th_column";

        # Output the results
        for (key in sum) {
            printf "%s%s%f%s%f\n", key, OFS, sum[key], OFS, max[key];
        }
    }
' | sort -V | column -t

# Human+Mouse trial
# sbatch -c 96 -p tux --job-name pggb-human-mouse --wrap "hostname; $PGGB -i /lizardfs/guarracino/pggb-paper/assemblies/grch38+mm39.fa.gz -o /lizardfs/guarracino/pggb-paper/graphs/grch38+mm39.p70 -p 70 -D /scratch"
# sbatch -c 96 -p tux -w tux09 --job-name pggb-human-mouse --wrap "hostname; $PGGB -i /lizardfs/guarracino/pggb-paper/assemblies/grch38+mm39.fa.gz -o /lizardfs/guarracino/pggb-paper/graphs/grch38+mm39.p70.k0 -p 70 -k 0 -P asm20 -D /scratch --input-paf /lizardfs/guarracino/pggb-paper/graphs/grch38+mm39.p70/grch38+mm39.fa.gz.707bd6e.alignments.wfmash.paf --resume"
```

Merge athaliana, tomato, and soy graphs:

```shell
for SPECIES in athaliana82 tomato23 soy37; do
  echo $SPECIES

  sed '1d' $PATH_PANGENOMES_TSV | sort -k 4n | grep $SPECIES | while read f p s n k REF; do
    seq 1 1 | while read c; do
      #echo $f $p $s $n $k $REF $c;

      SUFFIX=p$p.s$s.n$n.k${k}_$c
      echo $SUFFIX
    done
  done | sort | uniq | while read SUFFIX; do
    echo $SUFFIX
    ls $DIR_GRAPHS/$SPECIES*$SUFFIX/*final.og | sort -V | tr ' ' '\n' > $SUFFIX.tmp.txt
    $ODGI squeeze -f $SUFFIX.tmp.txt -o /scratch/$SPECIES.$SUFFIX.og -t 48 -P
    rm $SUFFIX.tmp.txt
    $ODGI view -i /scratch/$SPECIES.$SUFFIX.og -g > /scratch/$SPECIES.$SUFFIX.gfa
    mv /scratch/$SPECIES.$SUFFIX.og $DIR_GRAPHS/$SPECIES.$SUFFIX.og
    mv /scratch/$SPECIES.$SUFFIX.gfa $DIR_GRAPHS/$SPECIES.$SUFFIX.gfa
  done
done
```

## Minigraph Cactus

The first sample in `*.seqfile` is used to anchor the entire graph (https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/pangenome.md#sample-names).

Prepare the input:

```shell
cd $DIR_BASE

PATH_FASTA=/lizardfs/guarracino/pggb-paper/assemblies/athaliana/athaliana82.fa.gz
#PATH_FASTA=/lizardfs/guarracino/pggb-paper/assemblies/tomato/tomato23.fa.gz
#PATH_FASTA=/lizardfs/guarracino/pggb-paper/assemblies/soy/soy37.fa.gz
#sed '1d' $PATH_PANGENOMES_TSV | grep 'athaliana82\|tomato23\|soy37' -v | cut -f 1 -d ' ' | sort | uniq | while read PATH_FASTA; do
  DIR_PANGENOME="$(dirname $PATH_FASTA)"

  >&2 echo $DIR_PANGENOME $PATH_FASTA

  cut -f 1 $PATH_FASTA.fai | sort -k 2,2n | cut -f 1,2 -d '#' | grep ZH13 -v | sort | uniq | while read ACC; do
    ACC2=$(echo $ACC | tr '#' '_')

    if [[ "$ACC" == "chm13#1" ]] || [[ "$ACC" == "grch38#1" ]]; then
        ACC3="${ACC%#*}"
    else
        ACC3="$(echo $ACC | tr '.' '_' | tr '#' '.')"
    fi

    #samtools faidx $PATH_FASTA $(cut -f 1 $PATH_FASTA.fai | grep $ACC -w) | bgzip -l 9 -@ 48 > $DIR_PANGENOME/$ACC2.fasta.gz
    #samtools faidx $DIR_PANGENOME/$ACC2.fasta.gz
    echo $ACC3 $DIR_PANGENOME/$ACC2.fasta.gz | tr ' ' '\t'
  done #> $DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
done

# Fix references
reorder_seqfile() {
    local seqfile=$1
    local ref=$2
    local temp_file="temp_seqfile"

    grep "$ref" "$seqfile" > "$temp_file"
    grep "$ref" "$seqfile" -v >> "$temp_file"
    mv "$temp_file" "$seqfile"
}
reorder_seqfile "$DIR_BASE/assemblies/hsapiens/hsapiens90.chr6.masked.seqfile" "grch38"
reorder_seqfile "$DIR_BASE/assemblies/primates/primates16.hsa6.seqfile" "grch38"
reorder_seqfile "$DIR_BASE/assemblies/ecoli/ecoli50.seqfile" "GCA_000597845.1"
reorder_seqfile "$DIR_BASE/assemblies/ecoli/ecoli500.seqfile" "GCA_000597845.1"
reorder_seqfile "$DIR_BASE/assemblies/scerevisiae/scerevisiae142.seqfile" "SGDref"
reorder_seqfile "$DIR_BASE/assemblies/athaliana/athaliana82.seqfile" "GCA_028009825.2"
reorder_seqfile "$DIR_BASE/assemblies/tomato/tomato23.seqfile" "SL5"
reorder_seqfile "$DIR_BASE/assemblies/soy/soy37.seqfile" "ZH13" # Fix the ".seqfile" also manually

sed -e 's/\.M/\.2/g' -e 's/\.P/\.1/g' /lizardfs/guarracino/pggb-paper/assemblies/primates/primates16.hsa6.seqfile -i

mkdir -p $DIR_GRAPHS_MC

# On lambda01
export PATH=/lizardfs/guarracino/tools/cactus-bin-v2.7.0/bin:$PATH
source /lizardfs/guarracino/tools/cactus-bin-v2.7.0/venv-cactus-v2.7.0/bin/activate

cd ~
sed '1d' $PATH_PANGENOMES_TSV | grep 'athaliana82\|tomato23\|soy37' -v | grep 'mmusculus\|ecoli' | cut -f 1 -d ' ' | sort -r | uniq | while read PATH_FASTA; do
  DIR_PANGENOME="$(dirname $PATH_FASTA)"
  PATH_SEQFILE=$DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
  REF=$(head $PATH_SEQFILE -n 1 | cut -f 1)

  >&2 echo $PATH_SEQFILE $REF
  #sbatch -c 48 -p allnodes --job-name MC-$(basename $PATH_SEQFILE .seqfile) --wrap "hostname; cd /scratch; \time -v cactus-pangenome ./$(basename $PATH_FASTA .fa.gz)-js $PATH_SEQFILE --outDir $(basename $PATH_FASTA .fa.gz) --outName $(basename $PATH_FASTA .fa.gz) --reference $REF --vcf --giraffe --gfa --gbz --maxCores 48 --batchSystem single_machine --defaultCores 48 --vcfReference $REF --noSplit 2>&1 | tee $(basename $PATH_FASTA .fa.gz).MC.log; mv $(basename $PATH_FASTA .fa.gz)* $DIR_GRAPHS_MC"
  \time -v cactus-pangenome ./$(basename $PATH_FASTA .fa.gz)-js $PATH_SEQFILE --outDir $(basename $PATH_FASTA .fa.gz) --outName $(basename $PATH_FASTA .fa.gz) --reference $REF --vcf --giraffe --gfa --gbz --maxCores 48 --batchSystem single_machine --defaultCores 48 --vcfReference $REF --noSplit 2>&1 | tee $(basename $PATH_FASTA .fa.gz).MC.log; mv $(basename $PATH_FASTA .fa.gz)* $DIR_GRAPHS_MC
done


# MC will partition the assemblies by chromosome
cd ~

fmt="%C\n%Us user %Ss system %P cpu %es total %MKb max memory"
export PATH=/lizardfs/guarracino/tools/cactus-bin-v2.7.0/bin:$PATH
source /lizardfs/guarracino/tools/cactus-bin-v2.7.0/venv-cactus-v2.7.0/bin/activate

# athaliana82
PATH_FASTA=/lizardfs/guarracino/pggb-paper/assemblies/athaliana/athaliana82.fa.gz
DIR_PANGENOME="$(dirname $PATH_FASTA)"
#PATH_SEQFILE=$DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
#REF=$(head $PATH_SEQFILE -n 1 | cut -f 1)
ls $DIR_PANGENOME/*fasta.gz | while read FASTA; do ACC=$(basename $FASTA .fasta.gz | cut -f 1,2 -d '_' | tr '.' '_'); echo "$ACC\t$FASTA" >> Atha82.joblist; done 

/usr/bin/time -f $fmt cactus-pangenome ./js Atha82.joblist --outDir At82_MC --outName At82_MC --reference GCA_000001735_2 GCA_028009825_1 GCA_028009825_2 --vcf --giraffe --gfa --gbz --maxCores 224 --batchSystem single_machine --defaultCores 64 --vcfReference GCA_000001735_2 GCA_028009825_1 GCA_028009825_2
[2023-12-07T21:37:18+0000] [MainThread] [I] [toil.statsAndLogging] cactus-pangenome has finished after 89085.42339792382 seconds
2466690.39s user 69531.40s system 2846% cpu 89094.36s total 149417484Kb max memory
#\time -f $fmt cactus-pangenome ./$(basename $PATH_FASTA .fa.gz)-js $PATH_SEQFILE --outDir $(basename $PATH_FASTA .fa.gz) --outName $(basename $PATH_FASTA .fa.gz) --reference $REF --vcf --giraffe --gfa --gbz --maxCores 144 --batchSystem single_machine --defaultCores 144 --vcfReference $REF 2>&1 | tee $(basename $PATH_FASTA .fa.gz).MC.log; mv $(basename $PATH_FASTA .fa.gz)* $DIR_GRAPHS_MC


# athaliana82
PATH_FASTA=/lizardfs/guarracino/pggb-paper/assemblies/athaliana/athaliana82.fa.gz
DIR_PANGENOME="$(dirname $PATH_FASTA)"
PATH_SEQFILE=$DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
REF=$(head $PATH_SEQFILE -n 1 | cut -f 1)

\time -f $fmt cactus-pangenome ./$(basename $PATH_FASTA .fa.gz)-js $PATH_SEQFILE --outDir $(basename $PATH_FASTA .fa.gz) --outName $(basename $PATH_FASTA .fa.gz) --reference $REF --vcf --giraffe --gfa --gbz --maxCores 144 --batchSystem single_machine --defaultCores 144 --vcfReference $REF 2>&1 | tee $(basename $PATH_FASTA .fa.gz).MC.log; mv $(basename $PATH_FASTA .fa.gz)* $DIR_GRAPHS_MC

# tomato23
PATH_FASTA=/lizardfs/guarracino/pggb-paper/assemblies/tomato/tomato23.fa.gz
DIR_PANGENOME="$(dirname $PATH_FASTA)"
PATH_SEQFILE=$DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
REF=$(head $PATH_SEQFILE -n 1 | cut -f 1)

\time -f $fmt cactus-pangenome ./$(basename $PATH_FASTA .fa.gz)-js $PATH_SEQFILE --outDir $(basename $PATH_FASTA .fa.gz) --outName $(basename $PATH_FASTA .fa.gz) --reference $REF --vcf --giraffe --gfa --gbz --maxCores 144 --batchSystem single_machine --defaultCores 144 --vcfReference $REF 2>&1 | tee $(basename $PATH_FASTA .fa.gz).MC.log; mv $(basename $PATH_FASTA .fa.gz)* $DIR_GRAPHS_MC

# soy37
PATH_FASTA=/lizardfs/guarracino/pggb-paper/assemblies/soy/soy37.fa.gz
DIR_PANGENOME="$(dirname $PATH_FASTA)"
PATH_SEQFILE=$DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
REF=$(head $PATH_SEQFILE -n 1 | cut -f 1)

\time -f $fmt cactus-pangenome ./$(basename $PATH_FASTA .fa.gz)-js $PATH_SEQFILE --outDir $(basename $PATH_FASTA .fa.gz) --outName $(basename $PATH_FASTA .fa.gz) --reference $REF --vcf --giraffe --gfa --gbz --maxCores 144 --batchSystem single_machine --defaultCores 144 --vcfReference $REF 2>&1 | tee $(basename $PATH_FASTA .fa.gz).MC.log; mv $(basename $PATH_FASTA .fa.gz)* $DIR_GRAPHS_MC


cd $DIR_GRAPHS_MC

PATH_FASTA=$DIR_BASE/assemblies/ecoli/ecoli50.fa.gz
GFA_GZ=$DIR_GRAPHS_MC/ecoli50/ecoli50.gfa.gz

PATH_FASTA=$DIR_BASE/assemblies/ecoli/ecoli500.fa.gz
GFA_GZ=$DIR_GRAPHS_MC/ecoli500/ecoli500.gfa.gz

# Define an associative array with species names and file identifiers
declare -A species=( ["athaliana"]="athaliana82"
                     ["mmusculus"]="mmusculus36.chr19"
                     ["hsapiens"]="hsapiens90.chr6.masked"
                     ["primates"]="primates16.hsa6"
                     ["scerevisiae"]="scerevisiae142"
                     ["tomato"]="tomato23"
                     ["soy"]="soy37" )

# Iterate over the species array
for key in "${(k)species[@]}"; do
    echo $key

    # Construct file paths
    PATH_FASTA="$DIR_BASE/assemblies/$key/${species[$key]}.fa.gz"
    GFA_GZ="$DIR_GRAPHS_MC/${species[$key]}/${species[$key]}.gfa.gz"

    NAME=$(basename $GFA_GZ .gfa.gz)

    awk -v OFS='\t' '{print($1, "0", $2)}' "$PATH_FASTA.fai" | sed 's/\.\([^#]*#\)/_\1/'  | sed -e 's/#M#/#2#/g' -e 's/#P#/#1#/g' > $NAME.all.bed # replace only the 1st '.' (in the sample name) if it appears before the '#' character
    zgrep "^W" $GFA_GZ | cut -f 2,3,4,5,6 | awk -v OFS='\t' '{print($1"#"$2"#"$3,$4,$5)}' | sed 's/#0#/#1#/g' > $NAME.kept.bed

    (echo "#sequence length.bp removed.bp removed.ratio" | tr ' ' '\t'; join \
      <(cut -f 1,3 $NAME.all.bed | sort -k 1,1) \
      <(awk 'BEGIN{while(getline < "'$NAME.all.bed'") sum[$1]=0} {sum[$1] += $3 - $2} END{for(key in sum) print key, sum[key]}' <(bedtools subtract -a $NAME.all.bed -b $NAME.kept.bed) | sort -k 1,1) | awk -v OFS='\t' '
        {
            split($1, arr, "#");
            key = arr[1] "#" arr[2];
            length_sum[key] += $2;
            removed_sum[key] += $3
        }
        END {
            for (key in length_sum) {
                print(key, length_sum[key], removed_sum[key]);
            }
        }
    ' | awk -v OFS='\t' '{print($0,$3/$2)}') > $NAME.removed-seq.tsv
    awk -v OFS='\t' 'NR > 1 { sum_length += $2; sum_removed += $3 } END{print("all",sum_length,sum_removed,sum_removed/sum_length)}' $NAME.removed-seq.tsv >> $NAME.removed-seq.tsv
    rm $NAME.all.bed $NAME.kept.bed
done

grep all *tsv | sort -k 4,4nr | column -t

(echo sample length.bp removed.bp removed.ratio pangenome | tr ' ' '\t'; ls *removed-seq.tsv | while read f; do NAME=$(echo $f | sed 's/.removed-seq.tsv//g'); sed '1d' $f | grep all -v | awk -v OFS='\t' -v pangenome=$NAME '{print($0,pangenome)}'; done) > all.remove-seq.tsv
```

## PGGB with/without smoothxg

```shell
# H. sapiens
$SEQWISH -s $DIR_BASE/assemblies/hsapiens/hsapiens90.chr6.masked.fa.gz -p $DIR_GRAPHS/hsapiens90.chr6.masked_p98.s20000.n90.k47_1/hsapiens90.chr6.masked.fa.gz.445f03b.alignments.wfmash.paf -k 0 -f 0 -g $DIR_GRAPHS/hsapiens90.chr6.masked_p98.s20000.n90.k47_1/hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.gfa -B 10000000 -t 48 --temp-dir /scratch/ -P
$ODGI build -g $DIR_GRAPHS/hsapiens90.chr6.masked_p98.s20000.n90.k47_1/hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.gfa -o $DIR_GRAPHS/hsapiens90.chr6.masked_p98.s20000.n90.k47_1/hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.og -P

# A. thaliana
$SEQWISH -s $DIR_BASE/assemblies/athaliana/athaliana82.chr1.fa.gz -p $DIR_GRAPHS/athaliana82.chr1_p95.s10000.n82.k47_1/athaliana82.chr1.fa.gz.a13a25f.alignments.wfmash.paf -k 0 -f 0 -g $DIR_GRAPHS/athaliana82.chr1_p95.s10000.n82.k47_1/athaliana82.chr1_p95.s10000.n82.k47_1.seqwish.k0.gfa -B 10000000 -t 48 --temp-dir /scratch/ -P
$ODGI build -g $DIR_GRAPHS/athaliana82.chr1_p95.s10000.n82.k47_1/athaliana82.chr1_p95.s10000.n82.k47_1.seqwish.k0.gfa -o $DIR_GRAPHS/athaliana82.chr1_p95.s10000.n82.k47_1/athaliana82.chr1_p95.s10000.n82.k47_1.seqwish.k0.og -P


mkdir -p $DIR_BASE/smoothxg_evaluation
cd $DIR_BASE/smoothxg_evaluation

# H. sapiens
DIR_PANGENOME=$DIR_GRAPHS/hsapiens90.chr6.masked_p98.s20000.n90.k47_1
MAIN_GRAPH=$DIR_PANGENOME/hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.og
REF="chm13"


$ODGI paths -i $MAIN_GRAPH -L | cut -f 1,2 -d '#' | sort | uniq > $(basename $MAIN_GRAPH .og).prefixes.txt
$ODGI paths -i $MAIN_GRAPH -Ll | grep $REF | awk -v OFS='\t' '{print($1,$2-1,$3)}' | bedtools makewindows -b /dev/stdin -w 10000 > $(basename $MAIN_GRAPH .og).$REF.windows.10kbp.bed

ls $DIR_PANGENOME/*.og | while read GRAPH; do
  NAME=$(basename $GRAPH .og)
  echo $NAME

  # 1D
  $ODGI viz -i $GRAPH -o $NAME.merged_by_haplotype.png -x 1000 -y 500 -a 10 -M $(basename $MAIN_GRAPH .og).prefixes.txt -P

  # 2D
  $ODGI layout -i $GRAPH -N h -o $NAME.lay -T $NAME.lay.tsv -t 48 -P --temp-dir /scratch/
  $ODGI draw -i $GRAPH -c $NAME.lay -p $NAME.2D.png

  # Graph statistics
  $ODGI stats -i $GRAPH -S

  # Degree
  $ODGI degree -i $GRAPH -b $(basename $MAIN_GRAPH .og).$REF.windows.10kbp.bed -t 48 | bedtools sort > $NAME.degree.$REF.10kbp.tsv
  $ODGI degree -i $GRAPH -d > $NAME.degree.nodes.tsv

  # Depth
  $ODGI depth -i $GRAPH -b $(basename $MAIN_GRAPH .og).$REF.windows.10kbp.bed -t 48 | bedtools sort > $NAME.depth.$REF.10kbp.tsv
  $ODGI depth -i $GRAPH -d > $NAME.depth.nodes.tsv
done

[odgi::path_linear_sgd_layout] 2D path-guided SGD: 100.00% @ 3.25e+07 bp/s elapsed: 00:01:13:26 remain: 00:00:00:00 
[odgi::path_linear_sgd_layout] 2D path-guided SGD: 100.00% @ 3.09e+06 bp/s elapsed: 00:17:35:44 remain: 00:00:00:00

echo "#chrom start end value smoothxg metric" | tr ' ' '\t' > hsapiens90.chr6.masked_p98.s20000.n90.k47_1.degree-depth.$REF.10kbp.tsv
for metric in degree depth; do 
  cat \
    <(awk -v OFS='\t' -v metric=$metric '{print($0,"yes",metric)}' hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.$metric.$REF.10kbp.tsv) \
    <(awk -v OFS='\t' -v metric=$metric '{print($0,"no",metric)}' hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.$metric.$REF.10kbp.tsv) \
    >> hsapiens90.chr6.masked_p98.s20000.n90.k47_1.degree-depth.$REF.10kbp.tsv
done

#graph                                                                             length          nodes   edges   paths   steps
hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final                  212423296       4531659 6502016 1410    476956630
hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0   209420591       4370279 6148746 1410    651527277
#hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47  213486647       4053815 5636679 1410    584533180
```

C4:

```shell
GRAPH=chr6.C4.gfa
NAME=$(basename $graph .gfa)
PREFIX_REF=chm13

$ODGI extract -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.og -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -t 48 -P -O
$ODGI extract -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.gfa -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.og -t 48 -P -O
$ODGI extract -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.gfa -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -t 48 -P -O


$ODGI viz -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.png -m -B Spectral:4
$ODGI sort -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.og -p gYs -o - -P -t 10 --temp-dir /scratch | $ODGI viz -i - -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.sorted.png -m -B Spectral:4

$ODGI viz -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.png -m -B Spectral:4
$ODGI sort -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og  -p gYs -o - -P -t 10 --temp-dir /scratch | $ODGI viz -i - -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.sorted.png -m -B Spectral:4

# Parts
$ODGI viz -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.part.png -m -B Spectral:4 -p <($ODGI paths -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -L | grep 'chm13\|grch38\|HG00438\|HG00621\|HG01071') -M <($ODGI paths -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -L | grep 'chm13\|grch38\|HG00438\|HG00621\|HG01071' | cut -f 1,2 -d '#') -a 10 -x 1000 -y 150
$ODGI sort -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og  -p gYs -o - -P -t 10 --temp-dir /scratch | $ODGI viz -i - -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.part.sorted.png -m -B Spectral:4 -p <($ODGI paths -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -L | grep 'chm13\|grch38\|HG00438\|HG00621\|HG01071') -M <($ODGI paths -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -L | grep 'chm13\|grch38\|HG00438\|HG00621\|HG01071' | cut -f 1,2 -d '#') -a 10 -x 1000 -y 150

$ODGI viz -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.C4.png -m -B Spectral:4 -a 10 -x 1000 -y 150
$ODGI viz -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.C4.part.png -m -B Spectral:4 -p <($ODGI paths -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -L | grep 'chm13\|grch38\|HG00438\|HG00621\|HG01071') -M <($ODGI paths -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -L | grep 'chm13\|grch38\|HG00438\|HG00621\|HG01071' | cut -f 1,2 -d '#') -a 10 -x 1000 -y 150
```

Other tests:

```shell
REF="grch38"
PATH_GFA=$DIR_GRAPHS/hsapiens90.chr6.masked_p98.s20000.n90.k47_1/hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.gfa
PREFIX=$(echo $PATH_GFA | sed 's/.gfa//g')

vg deconstruct -P $REF -e -a -t $THREADS $PATH_GFA | bgzip -@ $THREADS -l 9 > $PREFIX.$REF.vcf.gz; vcfbub -l 0 -a 10000 --input $PREFIX.$REF.vcf.gz | vcfwave -I 1000 -t $THREADS > $PREFIX.$REF.decomposed.tmp.vcf; bcftools annotate -x INFO/TYPE $PREFIX.$REF.decomposed.tmp.vcf | awk '\$5 != "."' | bgzip -@ $THREADS -l 9 > $PREFIX.$REF.decomposed.vcf.gz; tabix $PREFIX.$REF.decomposed.vcf.gz; rm $PREFIX.$REF.decomposed.tmp.vcf;

VCF=$PREFIX.$REF.decomposed.vcf.gz
NAME="$(dirname $VCF | rev | cut -f 1 -d '/' | rev).k0"

VCF=$DIR_GRAPHS/hsapiens90.chr6.masked_p98.s20000.n90.k47_1/hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.grch38.decomposed.vcf.gz
NAME="$(dirname $VCF | rev | cut -f 1 -d '/' | rev)"

bash $DIR_BASE/scripts/evaluation.speciesN.sh \
    $VCF \
    grch38#1#chr6 \
    $NAME \
    $DIR_BASE/evaluations_vcfeval \
    $DIR_BASE/vcfs/hsapiens/grch38.fa \
    $DIR_BASE/vcfs/hsapiens/grch38.fa.sdf \
    $DIR_BASE/vcfs/hsapiens \
    .GRCh38_no_alt.deepvariant.pansn.vcf.gz \
    "":"" \
    ".easy":$DIR_BASE/vcfs/hsapiens/GRCh38_notinalldifficultregions.bed.gz \
    ".hard":$DIR_BASE/vcfs/hsapiens/GRCh38_alldifficultregions.bed.gz #\
    #".confident-yes":$DIR_BASE/vcfs/hsapiens/$SAMPLE.f1_assembly_v2.dip.confident-yes.pansn.bed.gz \
    #".lowmap-segdup":$DIR_BASE/vcfs/hsapiens/GRCh38_alllowmapandsegdupregions.bed.gz




$ODGI layout -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.lay --temp-dir /scratch -t 48 -P
$ODGI draw -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.og -c hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.lay -p hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.C4.2D.png

$ODGI layout -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.lay --temp-dir /scratch -t 48 -P
$ODGI draw -i hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.og -c hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.lay -p hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.C4.2D.png

ls *C4*.og | while read f; do echo $f; odgi stats -i $f -S; done

# quantify the runtime and memory too for odgi build/extract/sort?

cd $DIR_EVALUATIONS
GFA=/lizardfs/guarracino/pggb-paper/graphs/hsapiens90.chr6.masked_p98.s20000.n90.k47_1/hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k47.gfa
sbatch -c 48 -p workers -J hsapiens90.chr6.masked --wrap "hostname; cd /scratch; /lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh $GFA chm13 /lizardfs/guarracino/pggb-paper/evaluations/hsapiens90.chr6.masked_p98.s5000.n90.k79_1.seqwish.k47.gfa 48;"

GFA=/lizardfs/guarracino/pggb-paper/graphs/hsapiens90.chr6.masked_p98.s20000.n90.k47_1/hsapiens90.chr6.masked_p98.s20000.n90.k47_1.seqwish.k0.gfa
sbatch -c 48 -p workers -J hsapiens90.chr6.masked --wrap "hostname; cd /scratch; /lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh $GFA chm13 /lizardfs/guarracino/pggb-paper/evaluations/hsapiens90.chr6.masked_p98.s5000.n90.k79_1.seqwish.k0.gfa 48;"
```
