### All species

Variables:

```bash
RUN_PGGB=/home/guarracino/tools/pggb/pggb-4daed9c59ab09b75b03714274bf37e687b3a6db2
DIR_GRAPHS=/lizardfs/guarracino/pggb-paper/graphs
DIR_EVALUATIONS=/lizardfs/guarracino/pggb-paper/evaluations
PATH_PANGENOMES_TSV=/lizardfs/guarracino/pggb-paper/data/pangenomes.tsv
RUN_GFA_2_EVALUATION=/lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh
THREADS=48
```

Run `pggb`:

```bash
mkdir -p $DIR_GRAPHS
cd $DIR_GRAPHS

sed '1d' $PATH_PANGENOMES_TSV | while read f p s n k G POA O REF; do
  echo $f $p $s $n $k $G $POA $O $REF;

  out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')
  sbatch -c $t -p 386mem -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $RUN_PGGB -i $f -p $p -s $s -n $n -k $k -P $POA -O $O -G $G -t $THREADS -o $out; mv $out $DIR_GRAPHS/$out;"
done
```

Evaluation:

```bash
mkdir -p $DIR_EVALUATIONS
cd $DIR_EVALUATIONS

sed '1d' $PATH_PANGENOMES_TSV | while read f p s n k G POA O REF; do
  echo $f $p $s $n $k $G $POA $O $REF

  out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')
  PATH_GFA=$DIR_GRAPHS/$out/*.final.gfa
  echo "hostname; cd /scratch; $RUN_GFA_2_EVALUATION $PATH_GFA $REF $DIR_EVALUATIONS/$out $t;"
done
```

## NOTE

To remember that we are still doing this (both in PGGB and in GFA2EVALUATION) because of bugs in `vcfwave`:

```bash
# The TYPE info sometimes is wrong/missing
# There are variants without the ALT allele
bcftools annotate -x INFO/TYPE "$PATH_WAVED_VCF".gz  | awk '$5 != "."' | bgzip -@ THREADS -c > "$PATH_WAVED_FIXED_VCF".gz
```
