# Pangenome building

Variables:

```shell
DIR_BASE=/lizardfs/guarracino/pggb-paper
DIR_GRAPHS=/lizardfs/guarracino/pggb-paper/graphs
DIR_GRAPHS_MC=/lizardfs/guarracino/pggb-paper/graphs_mc
DIR_EVALUATIONS=/lizardfs/guarracino/pggb-paper/evaluations
PATH_PANGENOMES_TSV=/lizardfs/guarracino/pggb-paper/data/pangenomes.tsv

PGGB=/home/guarracino/tools/pggb/pggb-13482bd06359a7ad8e3d3e0dd6eb6d9399f26046
ODGI=/home/guarracino/tools/odgi/bin/odgi-861b1c04f5622c5bb916a161c1abe812c213f1a5
WFMASH=/home/guarracino/tools/wfmash/build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0
RUN_GFA_2_EVALUATION=/lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh

# guix install python-scikit-learn python-seaborn
PANACUS=/home/guarracino/tools/panacus/target/release/panacus-bd492f54c05367d0fc5a2c3fb9bf23260ac8379e
PANACUS_VISUALIZE=/home/guarracino/tools/panacus/scripts/panacus-visualize.py

THREADS=48
```

## PGGB

### Graph building

Run `pggb`:

```shell
mkdir -p $DIR_GRAPHS
cd $DIR_GRAPHS

sed '1d' $PATH_PANGENOMES_TSV | sort -k 4n | while read f p s n k G POA O REF; do
  seq 1 1 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;
    
    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    sbatch -c $THREADS -p 386mem -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $PGGB -i $f -p $p -s $s -n $n -k $k -P $POA -O $O -G $G -t $THREADS -o $out; mv $out $DIR_GRAPHS/$out;"
  done
done

cd $DIR_GRAPHS
# Script to parse statistics
(echo dataset tool runtime_s memory_kb replicate | tr ' ' '\t'; grep 'wfmash-\|seqwish-\|smoothxg-\|odgi-\|gfaffix-' */*log -A 1  | grep 'wfmash::align' -v | grep '^--$' -v | python3 ../scripts/pggb_stats.py | awk -v OFS='\t' '{print($0, substr($1, length($1), 1))}') > runtime+memory.stats.tsv

awk 'BEGIN { FS="\t"; OFS="\t" } 
  NR>1 { runtime[$1,$5]+=$3; memory[$1,$5]=(memory[$1,$5]>$4)?memory[$1,$5]:$4 } 
  END { 
      print "dataset", "replicate", "total_runtime_m", "max_memory_kb"; 
      for (key in runtime) { 
          split(key, indices, SUBSEP); 
          print indices[1], indices[2], runtime[key]/60, memory[key] 
      } 
  }' runtime+memory.stats.tsv > runtime+memory.stats.grouped.tsv

cat runtime+memory.stats.tsv | grep -v 'odgi-viz\|odgi-layout\|odgi-draw' | awk 'BEGIN { FS="\t"; OFS="\t" } 
  NR>1 { runtime[$1,$5]+=$3; memory[$1,$5]=(memory[$1,$5]>$4)?memory[$1,$5]:$4 } 
  END { 
      print "dataset", "replicate", "total_runtime_m", "max_memory_kb"; 
      for (key in runtime) { 
          split(key, indices, SUBSEP); 
          print indices[1], indices[2], runtime[key]/60, memory[key] 
      } 
  }' > runtime+memory.stats.grouped.no-viz.tsv

# Human+Mouse trial
sbatch -c 96 -p tux --job-name pggb-human-mouse --wrap "hostname; $PGGB -i /lizardfs/guarracino/pggb-paper/assemblies/grch38+mm39.fa.gz -o /lizardfs/guarracino/pggb-paper/graphs/grch38+mm39.p70 -p 70 -D /scratch"
sbatch -c 96 -p tux -w tux09 --job-name pggb-human-mouse --wrap "hostname; $PGGB -i /lizardfs/guarracino/pggb-paper/assemblies/grch38+mm39.fa.gz -o /lizardfs/guarracino/pggb-paper/graphs/grch38+mm39.p70.k0 -p 70 -k 0 -P asm20 -D /scratch --input-paf /lizardfs/guarracino/pggb-paper/graphs/grch38+mm39.p70/grch38+mm39.fa.gz.707bd6e.alignments.wfmash.paf --resume"
```

### Evaluation

SNVs evaluation with NUCMER:

```shell
mkdir -p $DIR_EVALUATIONS
cd $DIR_EVALUATIONS

sed '1d' $PATH_PANGENOMES_TSV | sort -k 4n | grep athaliana82 | grep 'chr1\|chr2\|chr3\|chr4' | while read f p s n k G POA O REF; do
  seq 1 1 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;

    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    PATH_GFA=$DIR_GRAPHS/$out/*.final.gfa
    PATH_GFA=$(eval echo $PATH_GFA) # force '*' expansion
    sbatch -c 96 -p tux -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $RUN_GFA_2_EVALUATION $PATH_GFA $REF $DIR_EVALUATIONS/$out 96;"
  done
done

cd $DIR_EVALUATIONS

for type in "" "waved."; do
  echo $type

  # All results
  ((echo -n replicate"\t"dataset"\t"; cat */statistics*.tsv | head -n 1 | tr -d '\n'; echo -e "\tpggb.evaluated"); ls */statistics.haplo.${type}tsv | while read f; do name=$(echo $f | cut -f 1 -d '/'); sed '1d' $f | awk -v name=$name -v OFS='\t' '{ print(name,$0) }' | awk -v OFS='\t' '{print(substr($1, length($1), 1),$0,$4+$5)}'; done) | sort -k 1,1n -k 2,2> nucmer-based_evaluation.${type}tsv

  # Average results
  # GCA_028009825.1 and GCA_028009825.2 are identical, so they create nan in the evaluation
  grep 'GCA_028009825.2' -v nucmer-based_evaluation.${type}tsv |
    awk '
    BEGIN { FS=OFS="\t" }
    NR>1 {
        count[($1 FS $2)]++
        for (i=4; i<=NF; i++) {  # Start from the 4th column
            sum[($1 FS $2 FS i)] += $i
        }
    }
    END {
        print "replicate", "dataset", "tp.baseline", "tp.call", "fp", "fn", "precision", "recall", "f1.score", "nucmer.tot", "pggb.tot", "nucmer.ratio", "pggb.ratio", "pggb.evaluated"
        for (key in count) {
            printf "%s", key
            for (i=4; i<=15; i++) {  # Adjust the indices here as well
                printf "%s%.6f", OFS, sum[key FS i] / count[key]
            }
            print ""
        }
    }' | sort -k 1,1n -k 2,2 > nucmer-based_evaluation.${type}mean.tsv
done
```

SNVs and INDELs evaluation: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXxx):
MAYBE ALSO SOY

tomato/athaliana/soy (for hsapiens, refer to the HPRC paper?)

```shell
# Call variants
cd $DIR_GRAPHS
sed '1d' $PATH_PANGENOMES_TSV | sort -k 4n | grep 'hsapiens\|tomato\|athaliana\|soy' | while read f p s n k G POA O REF; do
  seq 1 1 | while read c; do
    if [[ $f == *"hsapiens"* ]]; then
      REF="grch38" # hsapiens variants are called w.r.t. grch38
    fi
    echo $f $p $s $n $k $G $POA $O $REF $c;

    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    PATH_GFA=$DIR_GRAPHS/$out/*.final.gfa
    PATH_GFA=$(eval echo $PATH_GFA) # force '*' expansion
    PREFIX=$(echo $PATH_GFA | sed 's/.gfa//g')

    sbatch -c $THREADS -p allnodes --job-name vg-$out --wrap "hostname; cd $DIR_GRAPHS/$out; vg deconstruct -P $REF -e -a -t $THREADS $PATH_GFA | bgzip -@ $THREADS -l 9 > $PREFIX.$REF.vcf.gz; vcfbub -l 0 -a 10000 --input $PREFIX.$REF.vcf.gz | vcfwave -I 1000 -t $THREADS > $PREFIX.$REF.decomposed.tmp.vcf; bcftools annotate -x INFO/TYPE $PREFIX.$REF.decomposed.tmp.vcf | awk '\$5 != \".\"' | bgzip -@ $THREADS -l 9 > $PREFIX.$REF.decomposed.vcf.gz; tabix $PREFIX.$REF.decomposed.vcf.gz; rm $PREFIX.$REF.decomposed.tmp.vcf; zcat $PREFIX.$REF.decomposed.vcf.gz | bcftools view -a -Ou | bcftools norm -m - --threads $THREADS -Ou | bcftools sort -m 40G -T bcftools-sort.XXXXXX -Ou | bcftools norm -d exact -Ov --threads $THREADS | python3 /lizardfs/guarracino/rat_32samples/SV/fix_svtype_vcf.py | bgzip -@ $THREADS -l 9 > $PREFIX.$REF.decomposed.norm.vcf.gz"
  done
done

##################### FIX ATHALIANA NAMES
##################### FIX ATHALIANA NAMES
##################### FIX ATHALIANA NAMES
SED_CMD=""
while IFS= read -r line; do
    search=$(echo "$line" | awk '{print $1}')
    replace=$(echo "$line" | awk '{print $2}')
    SED_CMD+="s/$search/$replace/g;"
done < $DIR_BASE/data/athaliana82.acc2name.tsv
zcat xxx.vcf.gz | sed "$SED_CMD" "$TARGET_FILE" | bgzip -@ 48 -l 9 > x
mv x xxx.vcf.gz && tabix xxx.vcf.gz
##################### FIX ATHALIANA NAMES
##################### FIX ATHALIANA NAMES
##################### FIX ATHALIANA NAMES

# H. sapiens
cut -f 1 /lizardfs/erikg/HPRC/year1v2genbank/evaluation/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna -d ' ' | sed 's/>chr/>grch38#1#chr/g' > $DIR_BASE/vcfs/hsapiens/grch38.fa
samtools faidx /lizardfs/guarracino/pggb-paper/vcfs/hsapiens/grch38.fa
rtg format -o $DIR_BASE/vcfs/hsapiens/grch38.fa.sdf $DIR_BASE/vcfs/hsapiens/grch38.fa

# VCF files (from https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=submissions/B581EBA7-8BDE-4C7C-9DEA-78B99A051155--Yale_HPP_Year1_Variant_Calls/)
mkdir -p $DIR_BASE/vcfs/hsapiens
cd $DIR_BASE/vcfs/hsapiens
cut -f 2 $DIR_BASE/data/HPRC.deepvariant-vcf.urls.tsv | parallel -j 4 'wget -q {} && echo got {}'
ls $DIR_BASE/vcfs/hsapiens/*vcf.gz | while read VCF; do
  echo $VCF
  NAME=$(basename $VCF .vcf.gz)
  zcat $VCF | sed 's/chr/grch38#1#chr/g' | bgzip -@ 48 -l 9 > $NAME.pansn.vcf.gz
  rm $VCF
  tabix $NAME.pansn.vcf.gz
done

# Confident regions (from https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=working/HPRC/HG00438/assemblies/year1_freeze_assembly_v2/assembly_qc/dipcall_v0.2/)
cut -f 2 $DIR_BASE/data/HPRC.dipcall-confident-regions.urls.tsv | parallel -j 4 'wget -q {} && echo got {}'
ls $DIR_BASE/vcfs/hsapiens/*bed | while read BED; do
  echo $BED
  NAME=$(basename $BED .bed)
  cat $BED | sed 's/chr/grch38#1#chr/g' | sort -V | bgzip -@ 48 -l 9 > $NAME.confident-yes.pansn.bed.gz
  rm $BED
  tabix -p bed $NAME.confident-yes.pansn.bed.gz
done

# Stratification (from https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/genome-stratifications/v2.0/GRCh38/union/)
wget -c https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/genome-stratifications/v2.0/GRCh38/union/GRCh38_notinalldifficultregions.bed.gz # easy regions
wget -c https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/genome-stratifications/v2.0/GRCh38/union/GRCh38_alllowmapandsegdupregions.bed.gz # low map and SegDup
wget -c https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/genome-stratifications/v2.0/GRCh38/union/GRCh38_alldifficultregions.bed.gz # hard regions (low map and SegDup included)
for XXX in GRCh38_notinalldifficultregions.bed.gz GRCh38_alllowmapandsegdupregions.bed.gz GRCh38_alldifficultregions.bed.gz; do
  zcat $XXX | sed '1d' | sed 's/^chr/grch38#1#chr/g' | bgzip -@ 48 -l 9 > x
  mv x $XXX
  tabix -p bed $XXX
done

################################################################
################################################################
################################################################
################################################################
bash $DIR_BASE/scripts/evaluation.speciesN.sh \
  $DIR_BASE/graphs/hsapiens90.chr6.masked_p98.s10000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked.fa.gz.a8a102b.e34d4cd.80ba7c5.smooth.final.grch38.decomposed.vcf.gz \
  grch38#1#chr6 \
  hsapiens90.chr6.masked_p98.s10000.n90.k47.G700-900-1100.Pasm5.O0001_1 \
  $DIR_BASE/evaluations_vcfeval \
  $DIR_BASE/vcfs/hsapiens/grch38.fa \
  $DIR_BASE/vcfs/hsapiens/grch38.fa.sdf \
  $DIR_BASE/vcfs/hsapiens \
  .GRCh38_no_alt.deepvariant.pansn.vcf.gz \
  "":"" \
  ".easy":$DIR_BASE/vcfs/hsapiens/GRCh38_notinalldifficultregions.bed.gz \
  ".hard":$DIR_BASE/vcfs/hsapiens/GRCh38_alldifficultregions.bed.gz #\
  #".confident-yes":$DIR_BASE/vcfs/hsapiens/$SAMPLE.f1_assembly_v2.dip.confident-yes.pansn.bed.gz \
  #".lowmap-segdup":$DIR_BASE/vcfs/hsapiens/GRCh38_alllowmapandsegdupregions.bed.gz


# Tomato
mkdir -p $DIR_BASE/vcfs/tomato
cd $DIR_BASE/vcfs/tomato

zcat $DIR_BASE/assemblies/tomato/SL5.fa.gz > $DIR_BASE/vcfs/tomato/SL5.fa
samtools faidx $DIR_BASE/vcfs/tomato/SL5.fa
rtg format -o $DIR_BASE/vcfs/tomato/SL5.fa.sdf $DIR_BASE/vcfs/tomato/SL5.fa

# Transfer VCF files from Google Drive to /lizardfs/guarracino/pggb-paper/vcfs/tomato
# Fix header and apply PanSN
# Clean
rm $DIR_BASE/vcfs/tomato/*hifi.vcf.gz*

# Stratification
wget -c http://solomics.agis.org.cn/tomato/ftp/repeat/SL5.0.repeat.gff3.gz
zgrep -v "^#" SL5.0.repeat.gff3.gz | cut -f 1,4,5 | sort -k1,1V -k2,2n | bedtools merge -d 100 | sed 's/^/SL5#1#chr/g' | bedtools sort | bgzip -@ 48 -l 9 > SL5.hard.bed.gz # TE
bedtools subtract -a <(awk -v OFS='\t' '{print($1,"0",$2)}' $DIR_BASE/assemblies/tomato/SL5.fa.gz.fai) -b SL5.hard.bed.gz | bgzip -@ 48 -l 9 > SL5.easy.bed.gz

################################################################
################################################################
################################################################
################################################################
bash $DIR_BASE/scripts/evaluation.speciesN.sh \
  $DIR_BASE/graphs/tomato23.chr2_p90.s10000.n23.k17.G700-900-1100.Pasm5.O0001_1.SL5.decomposed.vcf.gz \
  SL5#1#chr2 \
  tomato23.chr2_p90.s10000.n23.k17.G700-900-1100.Pasm5.O0001_1 \
  $DIR_BASE/evaluations_vcfeval \
  $DIR_BASE/vcfs/tomato/SL5.fa \
  $DIR_BASE/vcfs/tomato/SL5.fa.sdf \
  $DIR_BASE/vcfs/tomato \
  .hifi.pansn.vcf.gz \
  "":"" \
  ".easy":$DIR_BASE/vcfs/tomato/SL5.easy.bed.gz \
  ".hard":$DIR_BASE/vcfs/tomato/SL5.hard.bed.gz

# A. thaliana
mkdir -p $DIR_BASE/vcfs/athaliana
cd $DIR_BASE/vcfs/athaliana

bcftools concat panCEN.dv.Chr*.vcf.gz | sed "s/Chr$i/SL5#1#chr$i/g" | bgzip -@ 48 -l 9 > All.hifi.pansn.vcf.gz && tabix All.hifi.pansn.vcf.gz
rm panCEN.dv.Chr*.vcf.gz 

zgrep '^#CHROM' All.hifi.pansn.vcf.gz -m 1 | cut -f 10- | tr '\t' '\n' | while read SAMPLE; do
  echo $SAMPLE
  bcftools view -s $SAMPLE -g hom -e 'GT="ref"' All.hifi.pansn.vcf.gz | sed 's/Chr/bgzip -@ 48 -l 9 > $SAMPLE.hifi.pansn.vcf.gz
  tabix $SAMPLE.hifi.pansn.vcf.gz
done
```

Compression:

```shell
cd $DIR_GRAPHS

(echo "dataset graph.len pangenome.len compression.ratio" | tr ' ' '\t'; ls $DIR_GRAPHS/*/*.final.og | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)"
  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)
  PREFIX=$DIR_PARENT/$NAME

  GRAPH_LEN=$($ODGI stats -i $GRAPH -S | tail -n 1 | cut -f 1)
  PANGENOME_LEN=$(cat $(grep Command $DIR_PARENT/*.log | cut -f 4 -d ' ').fai | awk '{sum+=$2}END{print(sum)}')
  echo $NAME $GRAPH_LEN $PANGENOME_LEN | awk -v OFS='\t' '{print($0,$3/$2)}'
done) > compression.stats.tsv
```

Openness:

```shell
ls $DIR_GRAPHS/*/*gfa | grep athaliana -v | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)"
  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)
  PREFIX=$DIR_PARENT/$NAME
  #NUM=$(echo $NAME | cut -f 1 -d '.' | grep -oP '\d+$')
  echo $PREFIX
  sbatch -c 48 -p tux --job-name panacus-$NAME --wrap "hostname; $PANACUS histgrowth $GRAPH -c bp -q 0,1,0.5,0.1 --groupby-sample -t 48 > $PREFIX.histgrowth.tsv; $PANACUS_VISUALIZE --estimate_growth_params $PREFIX.histgrowth.tsv -s 15 10 > $PREFIX.histgrowth.pdf"
done
```

Visualization by merging contigs by haplotype:

```shell
cd $DIR_BASE

ls $DIR_GRAPHS/*/*.final.og | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)"
  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)
  PREFIX=$DIR_PARENT/$NAME
  echo $NAME

  if [[ ! -s $PREFIX.prefixes.txt ]]; then
    $ODGI paths -i $GRAPH -L | cut -f 1,2 -d '#' | sort | uniq > $PREFIX.prefixes.txt
  fi
  
  sbatch -c 8 -p workers -J $NAME-N  --wrap "hostname; $ODGI viz -i $GRAPH -o $PREFIX.merged_by_haplotype.N.png  -x 1000 -y 500 -a 10 -N  -I Consensus_ -M $PREFIX.prefixes.txt"
  sbatch -c 8 -p workers -J $NAME-z  --wrap "hostname; $ODGI viz -i $GRAPH -o $PREFIX.merged_by_haplotype.z.png  -x 1000 -y 500 -a 10 -z  -I Consensus_ -M $PREFIX.prefixes.txt"
  sbatch -c 8 -p workers -J $NAME-du --wrap "hostname; $ODGI viz -i $GRAPH -o $PREFIX.merged_by_haplotype.du.png -x 1000 -y 500 -a 10 -du -I Consensus_ -M $PREFIX.prefixes.txt"
done
```

## Minigraph Cactus

The first sample in `*.seqfile` is used to anchor the entire graph (https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/pangenome.md#sample-names).

Prepare the input:

```shell
cd $DIR_BASE

sed '1d' $PATH_PANGENOMES_TSV | grep athaliana82 | cut -f 1 -d ' ' | sort | uniq | while read PATH_FASTA; do
  DIR_PANGENOME="$(dirname $PATH_FASTA)"

  >&2 echo $DIR_PANGENOME $PATH_FASTA

  cut -f 1 $PATH_FASTA.fai | sort -k 2,2n | cut -f 1,2 -d '#' | sort | uniq | while read ACC; do
    ACC2=$(echo $ACC | tr '#' '_')

    if [[ "$ACC" == "chm13#1" ]] || [[ "$ACC" == "grch38#1" ]]; then
        ACC3="${ACC%#*}"
    else
        ACC3="$(echo $ACC | tr '#' '.')"
    fi

    #samtools faidx $PATH_FASTA $(cut -f 1 $PATH_FASTA.fai | grep $ACC) | bgzip -l 9 -@ 48 > $DIR_PANGENOME/$ACC2.fasta.gz
    #samtools faidx $DIR_PANGENOME/$ACC2.fasta.gz
    echo $ACC3 $DIR_PANGENOME/$ACC2.fasta.gz | tr ' ' '\t'
  done > $DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
done

# MC will partition the assemblies by chromosome
cd $DIR_BASE/assemblies/athaliana
mv athaliana82.chr1.seqfile athaliana82.seqfile
rm athaliana82.chr2.seqfile athaliana82.chr3.seqfile athaliana82.chr4.seqfile athaliana82.chr5.seqfile
cd $DIR_BASE

# Fix references
reorder_seqfile() {
    local seqfile=$1
    local ref=$2
    local temp_file="temp_seqfile"

    grep "$ref" "$seqfile" > "$temp_file"
    grep "$ref" "$seqfile" -v >> "$temp_file"
    mv "$temp_file" "$seqfile"
}
reorder_seqfile "$DIR_BASE/assemblies/primates/primates16.hsa6.seqfile" "grch38"
reorder_seqfile "$DIR_BASE/assemblies/tomato/tomato23.chr2.seqfile" "SL5"
reorder_seqfile "$DIR_BASE/assemblies/ecoli/ecoli50.seqfile" "GCA_000597845.1"
reorder_seqfile "$DIR_BASE/assemblies/ecoli/ecoli500.seqfile" "GCA_000597845.1"
reorder_seqfile "$DIR_BASE/assemblies/scerevisiae/scerevisiae142.seqfile" "SGDref"
reorder_seqfile "$DIR_BASE/assemblies/athaliana/athaliana82.seqfile" "GCA_028009825.1"


# On lambda01
mkdir -p $DIR_GRAPHS_MC
cd $DIR_GRAPHS_MC

fmt="%C\n%Us user %Ss system %P cpu %es total %MKb max memory"
export PATH=/lizardfs/guarracino/tools/cactus-bin-v2.7.0/bin:$PATH
source /lizardfs/guarracino/tools/cactus-bin-v2.7.0/venv-cactus-v2.7.0/bin/activate

sed '1d' $PATH_PANGENOMES_TSV | grep athaliana82 -v | cut -f 1 -d ' ' | sort | uniq | while read PATH_FASTA; do
  DIR_PANGENOME="$(dirname $PATH_FASTA)"
  PATH_SEQFILE=$DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
  REF=$(head $PATH_SEQFILE -n 1 | cut -f 1)

  >&2 echo $PATH_SEQFILE $REF
  echo sbatch -c 48 -p 386mem --job-name MC-$(basename $PATH_SEQFILE .seqfile) --wrap "hostname; cd /scratch; \time -f $fmt cactus-pangenome ./$(basename $PATH_FASTA .fa.gz)-js $PATH_SEQFILE --outDir $(basename $PATH_FASTA .fa.gz) --outName $(basename $PATH_FASTA .fa.gz) --reference $REF --vcf --giraffe --gfa --gbz --maxCores 48 --batchSystem single_machine --defaultCores 48 --vcfReference $REF --noSplit 2>&1 | tee $(basename $PATH_FASTA .fa.gz).MC.log; mv $(basename $PATH_FASTA .fa.gz)* $DIR_GRAPHS_MC"
done









DIR_PANGENOME=/lizardfs/guarracino/pggb-paper/assemblies/athaliana
cut -f 1 $DIR_PANGENOME/athaliana82.fa.gz.fai| cut -f 1 -d '#' | sort | uniq | while read ACC; do echo $ACC; samtools faidx $DIR_PANGENOME/athaliana82.fa.gz $(cut -f 1 $DIR_PANGENOME/athaliana82.fa.gz.fai | grep $ACC) | bgzip -l 9 -@ 48 > $DIR_PANGENOME/$ACC.fasta.gz; samtools faidx $DIR_PANGENOME/$ACC.fasta.gz; done
rm Atha82.joblist
ls $DIR_PANGENOME/*fasta.gz | while read FASTA; do ACC=$(basename $FASTA .fasta.gz | cut -f 1,2 -d '_' | tr '.' '_'); echo "$ACC\t$FASTA" >> Atha82.joblist; done 

#!/bin/bash
fmt="%C\n%Us user %Ss system %P cpu %es total %MKb max memory"
export PATH=/lizardfs/guarracino/tools/cactus-bin-v2.7.0/bin:$PATH
source /lizardfs/guarracino/tools/cactus-bin-v2.7.0/venv-cactus-v2.7.0/bin/activate
/usr/bin/time -f $fmt cactus-pangenome ./js Atha82.joblist --outDir At82_MC --outName At82_MC --reference GCA_000001735_2 GCA_028009825_1 GCA_028009825_2 --vcf --giraffe --gfa --gbz --maxCores 224 --batchSystem single_machine --defaultCores 64 --vcfReference GCA_000001735_2 GCA_028009825_1 GCA_028009825_2
[2023-12-07T21:37:18+0000] [MainThread] [I] [toil.statsAndLogging] cactus-pangenome has finished after 89085.42339792382 seconds
2466690.39s user 69531.40s system 2846% cpu 89094.36s total 149417484Kb max memory

DIR_PANGENOME=/lizardfs/guarracino/pggb-paper/assemblies/primates
cut -f 1 $DIR_PANGENOME/primates16.hsa6.fa.gz.fai| cut -f 1 -d '#' | sort | uniq | while read ACC; do echo $ACC; samtools faidx $DIR_PANGENOME/primates16.hsa6.fa.gz $(cut -f 1 $DIR_PANGENOME/primates16.hsa6.fa.gz.fai | grep $ACC) | bgzip -l 9 -@ 48 > $DIR_PANGENOME/$ACC.hsa6.fasta.gz; samtools faidx $DIR_PANGENOME/$ACC.hsa6.fasta.gz; done
rm primates16.hsa6.joblist
ls $DIR_PANGENOME/*.hsa6.fasta.gz | while read FASTA; do ACC=$(basename $FASTA .fasta.gz | cut -f 1,2 -d '_' | tr '.' '_'); echo "$ACC\t$FASTA" >> primates16.hsa6.joblist; done 
[2023-12-07T08:28:49+0000] [MainThread] [I] [toil.statsAndLogging] cactus-pangenome has finished after 14534.32546418719 seconds
39138.20s user 7334.40s system 319% cpu 14536.33s total 57772840Kb max memory  

#!/bin/bash
fmt="%C\n%Us user %Ss system %P cpu %es total %MKb max memory"
export PATH=/lizardfs/guarracino/tools/cactus-bin-v2.7.0/bin:$PATH
source /lizardfs/guarracino/tools/cactus-bin-v2.7.0/venv-cactus-v2.7.0/bin/activate
/usr/bin/time -f $fmt cactus-pangenome ./js-primates16.hsa6 primates16.hsa6.joblist --outDir primates16.hsa6_MC --outName primates16.hsa6_MC --reference grch38_hsa6 --vcf --giraffe --gfa --gbz --maxCores 32 --batchSystem single_machine --defaultCores 8 --vcfReference grch38_hsa6
/usr/bin/time -f $fmt cactus-pangenome ./js-primates16.hsa6-no-split primates16.hsa6.joblist --outDir primates16.hsa6_MC-no-split --outName primates16.hsa6_MC-no-split --reference grch38_hsa6 --vcf --giraffe --gfa --gbz --maxCores 32 --batchSystem single_machine --defaultCores 8 --vcfReference grch38_hsa6 --noSplit
[2023-12-07T19:34:41+0000] [MainThread] [I] [toil.statsAndLogging] cactus-pangenome has finished after 14368.12507063197 seconds
36943.98s user 7407.71s system 308% cpu 14372.01s total 57775112Kb max memory
```

## PGGB with/without smoothxg

```shell
/home/guarracino/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9 -s /lizardfs/guarracino/pggb-paper/assemblies/hsapiens/hsapiens90.chr6.masked.fa.gz -p hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked.fa.gz.445f03b.alignments.wfmash.paf -k 47 -f 0 -g hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.gfa -B 10000000 -t 32 --temp-dir /scratch/ -P

/home/guarracino/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9 -s /lizardfs/guarracino/pggb-paper/assemblies/hsapiens/hsapiens90.chr6.masked.fa.gz -p hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked.fa.gz.445f03b.alignments.wfmash.paf -k 0 -f 0 -g hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.gfa -B 10000000 -t 32 --temp-dir /scratch/ -P


ls *gfa | grep '\.k' | while read GRAPH; do echo $GRAPH; odgi build -g $GRAPH -o $(basename $GRAPH .gfa).og -t 32; done  

ls *og | while read GRAPH; do echo $GRAPH; odgi stats -i $GRAPH -S; done       

#graph                                                                             length          nodes   edges   paths   steps
hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final                  212423296       4531659 6502016 1410    476956630
hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0   209420591       4370279 6148746 1410    651527277
hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47  213486647       4053815 5636679 1410    584533180


GRAPH=chr6.C4.gfa
NAME=$(basename $graph .gfa)
PREFIX_REF=chm13

odgi paths -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.og -Ll | grep chm13 | awk -v OFS='\t' '{print($1,$2-1,$3)}' | bedtools makewindows -b /dev/stdin -w 10000 > hsapiens90.chr6.masked.chm13.windows.10kbp.bed

odgi depth -i $GRAPH -b hsapiens90.chr6.masked.chm13.windows.10kbp.bed | bedtools sort > xxx
odgi degree -i $GRAPH -b hsapiens90.chr6.masked.chm13.windows.10kbp.bed | bedtools sort > xxx
odgi degree -i $GRAPH -d > xxx.tsv









$ODGI extract -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.og -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -t 48 -P -O
$ODGI extract -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.gfa -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -t 48 -P -O
$ODGI extract -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.gfa -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -t 48 -P -O


$ODGI viz -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.unsorted.png -m -B Spectral:4
$ODGI sort -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -p gYs -o - -P -t 10 --temp-dir /scratch | $ODGI viz -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.png -m -B Spectral:4

$ODGI viz -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.unsorted.png -m -B Spectral:4
$ODGI sort -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -p gYs -o - -P -t 10 --temp-dir /scratch | $ODGI viz -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.png -m -B Spectral:4


$ODGI layout -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.lay --temp-dir /scratch -t 48 -P
$ODGI draw -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -c hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.lay -p hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.2D.png

$ODGI layout -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.lay --temp-dir /scratch -t 48 -P
$ODGI draw -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -c hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.lay -p hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.2D.png

ls *C4*.og | while read f; do echo $f; odgi stats -i $f -S; done

# quantify the runtime and memory too for odgi build/extract/sort?
# we would need to find a locus where -k 47 lead to underalignment and one where -k 0 lead to overalignment



cd $DIR_EVALUATIONS
GFA=/lizardfs/guarracino/pggb-paper/graphs/hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.gfa
sbatch -c 48 -p workers -J hsapiens90.chr6.masked --wrap "hostname; cd /scratch; /lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh $GFA chm13 /lizardfs/guarracino/pggb-paper/evaluations/hsapiens90.chr6.masked_p98.s5000.n90.k79.G700-900-1100.Pasm5.O0001_1.seqwish.k47.gfa 48;"

GFA=/lizardfs/guarracino/pggb-paper/graphs/hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.gfa
sbatch -c 48 -p workers -J hsapiens90.chr6.masked --wrap "hostname; cd /scratch; /lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh $GFA chm13 /lizardfs/guarracino/pggb-paper/evaluations/hsapiens90.chr6.masked_p98.s5000.n90.k79.G700-900-1100.Pasm5.O0001_1.seqwish.k0.gfa 48;"
```
