# Pangenome building

## All species

Variables:

```shell
RUN_PGGB=/home/guarracino/tools/pggb/pggb-b3c55023313c28d89281d2e0ca0f10d25685d4e7
DIR_GRAPHS=/lizardfs/guarracino/pggb-paper/graphs
DIR_EVALUATIONS=/lizardfs/guarracino/pggb-paper/evaluations
PATH_PANGENOMES_TSV=/lizardfs/guarracino/pggb-paper/data/pangenomes.tsv
RUN_GFA_2_EVALUATION=/lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh
THREADS=48
```

Run `pggb`:

```shell
mkdir -p $DIR_GRAPHS
cd $DIR_GRAPHS

sed '1d' $PATH_PANGENOMES_TSV | grep rat -v | sort -k 4n | while read f p s n k G POA O REF; do
  seq 3 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;
    
    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    echo sbatch -c $THREADS -p 386mem -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $RUN_PGGB -i $f -p $p -s $s -n $n -k $k -P $POA -O $O -G $G -t $THREADS -o $out; mv $out $DIR_GRAPHS/$out;"
  done
done

# Script to parse statistics
grep 'wfmash-\|seqwish-\|smoothxg-\|odgi-\|gfaffix[[:space:]]' */*log -A 1  | grep 'wfmash::align' -v | grep '^--$' -v | python3 ../scripts/pggb_stats.py | column -t

#Old
#grep 'wfmash-\|seqwish-\|smoothxg-\|odgi-' hsapiens90.chr6.masked_p98.s10000.n90.k79.G700-900-1100.Pasm5.O0001/*log -A 1  | grep 'wfmash::align' -v | grep '^--$' -v | cut -f 1,2 -d ' ' | rev | cut -f 1 -d '/' | rev | sed 's/-[^ ]*//g' | sed 's/s user//g' | paste - - | sed 's/ \([[:alpha:]]\)/_\1/g' | column -t
```

Evaluation:

```shell
mkdir -p $DIR_EVALUATIONS
cd $DIR_EVALUATIONS

sed '1d' $PATH_PANGENOMES_TSV | grep rat -v | sort -k 4n | grep scerevisiae142 | while read f p s n k G POA O REF; do
  seq 3 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;

    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g').xauto_$c
    PATH_GFA=$DIR_GRAPHS/$out/*.final.gfa
    PATH_GFA=$(eval echo $PATH_GFA) # force '*' expansion
    sbatch -c $THREADS -p workers -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $RUN_GFA_2_EVALUATION $PATH_GFA $REF $DIR_EVALUATIONS/$out $THREADS;"
  done
done

# Risky verbose way
ls */*log | while read f; do x=$(echo $f | cut -f 1 -d '.'); grep 'max memory' $f | head -n 4 | cut -f 7 -d ' ' | sed 's/s//g' | awk '{print($0/60/60)}' | paste - - - - | awk '{print($1,"wfmash-mapping", $2, "wfmash-alignment",$3,"seqwish",$4,"smoothxg")}' | tr ' ' '\n' | paste - - | awk -v set=$x '{print($0,set)}' | column -t; done | column -t

( cat scerevisiae142_p95.s5000.n142.k47.G700-900-1100.Pasm5.O0001/statistics.haplo.tsv | head -n 1; ls scerevisiae142_p95.s5000.n142.k47.G700-900-1100.Pasm5.O0001*/statistics.haplo*tsv | while read f; do awk '{ for (i=1; i<=NF; i++) {sum[i] += $i} } END { for (i=1; i<=NF; i++) printf("%.6f ", sum[i]/NR); printf("\n") }' $f; done) | column -t

(cat */statistics*.tsv | head -n 1; ls */statistics*.tsv | while read f; do name=$(echo $f | cut -f 1 -d '/'); awk -v name=$name '{ for (i=1; i<=NF; i++) {sum[i] += $i} } END { printf(name,""); for (i=1; i<=NF; i++) printf("%.6f ", sum[i]/NR); printf("\n") }' $f; done) | column -t | less -S
```

### Primates

```shell
seq 6 7 | while read i; do
  f=/lizardfs/guarracino/tree_of_life_alignment/partitioning/primates17.chr$i.fa.gz
  out=primates17.chr${i}_p90.s5000.n4.k79.G700-900-1100.Pasm5.O0001_1

  sbatch -c 48 -p 386mem -J primates17.chr$i --wrap "hostname; cd /scratch; /home/guarracino/tools/pggb/pggb-b3c55023313c28d89281d2e0ca0f10d25685d4e7 -i $f -p 90 -s 5000 -n 17 -k 79 -P asm5 -O 0.001 -G 700,900,1100 -t 48 -o $out; mv $out /lizardfs/guarracino/pggb-paper/graphs/$out;"
done
```


## NOTES

To remember that we are still doing this (both in PGGB and in GFA2EVALUATION) because of bugs in `vcfwave`:

```shell
# The TYPE info sometimes is wrong/missing
# There are variants without the ALT allele
bcftools annotate -x INFO/TYPE "$PATH_WAVED_VCF".gz  | awk '$5 != "."' | bgzip -@ THREADS -c > "$PATH_WAVED_FIXED_VCF".gz
```
