# Pangenome building

## All species

Variables:

```shell
DIR_BASE=/lizardfs/guarracino/pggb-paper
DIR_GRAPHS=/lizardfs/guarracino/pggb-paper/graphs
DIR_GRAPHS_MC=/lizardfs/guarracino/pggb-paper/graphs_mc
DIR_EVALUATIONS=/lizardfs/guarracino/pggb-paper/evaluations
PATH_PANGENOMES_TSV=/lizardfs/guarracino/pggb-paper/data/pangenomes.tsv

PGGB=/home/guarracino/tools/pggb/pggb-13482bd06359a7ad8e3d3e0dd6eb6d9399f26046
ODGI=/home/guarracino/tools/odgi/bin/odgi-861b1c04f5622c5bb916a161c1abe812c213f1a5
WFMASH=/home/guarracino/tools/wfmash/build/bin/wfmash-0b191bb84ffdfd257354c1aa82a7f1e13dc536d0
RUN_GFA_2_EVALUATION=/lizardfs/guarracino/pggb-paper/scripts/gfa2evaluation.sh

THREADS=48
```

Run `pggb`:

```shell
mkdir -p $DIR_GRAPHS
cd $DIR_GRAPHS

sed '1d' $PATH_PANGENOMES_TSV | sort -k 4n | while read f p s n k G POA O REF; do
  seq 1 1 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;
    
    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    sbatch -c $THREADS -p 386mem -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $PGGB -i $f -p $p -s $s -n $n -k $k -P $POA -O $O -G $G -t $THREADS -o $out; mv $out $DIR_GRAPHS/$out;"
  done
done

cd $DIR_GRAPHS
# Script to parse statistics
(echo dataset tool runtime_s memory_kb replicate | tr ' ' '\t'; grep 'wfmash-\|seqwish-\|smoothxg-\|odgi-\|gfaffix-' */*log -A 1  | grep 'wfmash::align' -v | grep '^--$' -v | python3 ../scripts/pggb_stats.py | awk -v OFS='\t' '{print($0, substr($1, length($1), 1))}') > runtime+memory.stats.tsv

awk 'BEGIN { FS="\t"; OFS="\t" } 
  NR>1 { runtime[$1,$5]+=$3; memory[$1,$5]=(memory[$1,$5]>$4)?memory[$1,$5]:$4 } 
  END { 
      print "dataset", "replicate", "total_runtime_m", "max_memory_kb"; 
      for (key in runtime) { 
          split(key, indices, SUBSEP); 
          print indices[1], indices[2], runtime[key]/60, memory[key] 
      } 
  }' runtime+memory.stats.tsv > runtime+memory.stats.grouped.tsv

cat runtime+memory.stats.tsv | grep -v 'odgi-viz\|odgi-layout\|odgi-draw' | awk 'BEGIN { FS="\t"; OFS="\t" } 
  NR>1 { runtime[$1,$5]+=$3; memory[$1,$5]=(memory[$1,$5]>$4)?memory[$1,$5]:$4 } 
  END { 
      print "dataset", "replicate", "total_runtime_m", "max_memory_kb"; 
      for (key in runtime) { 
          split(key, indices, SUBSEP); 
          print indices[1], indices[2], runtime[key]/60, memory[key] 
      } 
  }' > runtime+memory.stats.grouped.no-viz.tsv

#Old
#grep 'wfmash-\|seqwish-\|smoothxg-\|odgi-' hsapiens90.chr6.masked_p98.s10000.n90.k79.G700-900-1100.Pasm5.O0001/*log -A 1  | grep 'wfmash::align' -v | grep '^--$' -v | cut -f 1,2 -d ' ' | rev | cut -f 1 -d '/' | rev | sed 's/-[^ ]*//g' | sed 's/s user//g' | paste - - | sed 's/ \([[:alpha:]]\)/_\1/g' | column -t

# Risky verbose way
#ls */*log | while read f; do x=$(echo $f | cut -f 1 -d '.'); grep 'max memory' $f | head -n 4 | cut -f 7 -d ' ' | sed 's/s//g' | awk '{print($0/60/60)}' | paste - - - - | awk '{print($1,"wfmash-mapping", $2, #"wfmash-alignment",$3,"seqwish",$4,"smoothxg")}' | tr ' ' '\n' | paste - - | awk -v set=$x '{print($0,set)}' | column -t; done | column -t
```

Evaluation:

```shell
mkdir -p $DIR_EVALUATIONS
cd $DIR_EVALUATIONS

# To RUN THE LAST ECOLI500 EVALUATION WHEN THE GFA IS AVAILABLE
sed '1d' $PATH_PANGENOMES_TSV | grep 'ecoli500' | head -n 4 | tail -n 1 | sort -k 4n | while read f p s n k G POA O REF; do
#sed '1d' $PATH_PANGENOMES_TSV | grep 'athaliana' | sort -k 4n | while read f p s n k G POA O REF; do
  seq 1 1 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;

    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c.xauto
    #out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    PATH_GFA=$DIR_GRAPHS/$out/*.final.gfa
    PATH_GFA=$(eval echo $PATH_GFA) # force '*' expansion
    sbatch -c $THREADS -p workers -J $(basename "$f" .fa.gz) --wrap "hostname; cd /scratch; $RUN_GFA_2_EVALUATION $PATH_GFA $REF $DIR_EVALUATIONS/$out $THREADS;"
  done
done

cd $DIR_EVALUATIONS

for type in "" "waved."; do
  echo $type

  # All results
  ((echo -n replicate"\t"dataset"\t"; cat */statistics*.tsv | head -n 1 | tr -d '\n'; echo -e "\tpggb.evaluated"); ls */statistics.haplo.${type}tsv | while read f; do name=$(echo $f | cut -f 1 -d '/'); sed '1d' $f | awk -v name=$name -v OFS='\t' '{ print(name,$0) }' | awk -v OFS='\t' '{print(substr($1, length($1), 1),$0,$4+$5)}'; done) | sort -k 1,1n -k 2,2> nucmer-based_evaluation.${type}tsv

  # Average results
  awk '
  BEGIN { FS=OFS="\t" }
  NR>1 {
      count[($1 FS $2)]++
      for (i=4; i<=NF; i++) {  # Start from the 4th column
          sum[($1 FS $2 FS i)] += $i
      }
  }
  END {
      print "replicate", "dataset", "tp.baseline", "tp.call", "fp", "fn", "precision", "recall", "f1.score", "nucmer.tot", "pggb.tot", "nucmer.ratio", "pggb.ratio", "pggb.evaluated"
      for (key in count) {
          printf "%s", key
          for (i=4; i<=15; i++) {  # Adjust the indices here as well
              printf "%s%.6f", OFS, sum[key FS i] / count[key]
          }
          print ""
      }
  }' nucmer-based_evaluation.${type}tsv | sort -k 1,1n -k 2,2 > nucmer-based_evaluation.${type}mean.tsv
done
```

Compression:

```shell
cd $DIR_GRAPHS

(echo "dataset graph.len pangenome.len compression.ratio" | tr ' ' '\t'; ls $DIR_GRAPHS/*/*.final.og | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)"
  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)
  PREFIX=$DIR_PARENT/$NAME

  GRAPH_LEN=$($ODGI stats -i $GRAPH -S | tail -n 1 | cut -f 1)
  PANGENOME_LEN=$(cat $(grep Command $DIR_PARENT/*.log | cut -f 4 -d ' ').fai | awk '{sum+=$2}END{print(sum)}')
  echo $NAME $GRAPH_LEN $PANGENOME_LEN | awk -v OFS='\t' '{print($0,$3/$2)}'
done) > compression.stats.tsv
```

Openness (INCOMPLETE):

```shell
PANACUS=/home/guarracino/tools/panacus/target/release/panacus-bd492f54c05367d0fc5a2c3fb9bf23260ac8379e
PANACUS_VISUALIZE=/home/guarracino/tools/panacus/scripts/panacus-visualize.py
GFA=/lizardfs/guarracino/pggb-paper/graphs/tomato23.chr2_p95.s10000.n23.k17.G700-900-1100.Pasm5.O0001_1/tomato23.chr2.fa.gz.a13a25f.ebe675d.eff8dac.smooth.final.gfa
NAME=$(basename $GFA .gfa)
$PANACUS histgrowth $GFA -c bp -q 0,1,0.5,0.1 -t 48 > $NAME.histgrowth.tsv

# guix install python-scikit-learn python-seaborn
$PANACUS_VISUALIZE -e -l "upper left" $NAME.histgrowth.tsv > $NAME.histgrowth.pdf
```

Visualization by merging contigs by haplotype:

```shell
cd $DIR_BASE

ls $DIR_GRAPHS/*/*.final.og | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)"
  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)
  PREFIX=$DIR_PARENT/$NAME
  echo $NAME

  if [[ ! -s $PREFIX.prefixes.txt ]]; then
    $ODGI paths -i $GRAPH -L | cut -f 1,2 -d '#' | sort | uniq > $PREFIX.prefixes.txt
  fi
  
  sbatch -c 8 -p workers -J $NAME-N  --wrap "hostname; $ODGI viz -i $GRAPH -o $PREFIX.merged_by_haplotype.N.png  -x 1000 -y 500 -a 10 -N  -I Consensus_ -M $PREFIX.prefixes.txt"
  sbatch -c 8 -p workers -J $NAME-z  --wrap "hostname; $ODGI viz -i $GRAPH -o $PREFIX.merged_by_haplotype.z.png  -x 1000 -y 500 -a 10 -z  -I Consensus_ -M $PREFIX.prefixes.txt"
  sbatch -c 8 -p workers -J $NAME-du --wrap "hostname; $ODGI viz -i $GRAPH -o $PREFIX.merged_by_haplotype.du.png -x 1000 -y 500 -a 10 -du -I Consensus_ -M $PREFIX.prefixes.txt"
done
```

### Primates

Layout:

```shell
cd $DIR_BASE

ls $DIR_GRAPHS/primates16.hsa6_p9*/*.og | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)"
  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)

  PREFIX=$DIR_PARENT/$NAME.x300.G20.I10000.l1000.Nh
  sbatch -c 48 -p workers -J $NAME --wrap "hostname; $ODGI layout -i $GRAPH -x 300 -G 20 -I 10000 -l 1000 -N h -o $PREFIX.lay -T $PREFIX.lay.tsv -t 48 -P --temp-dir /scratch/; $ODGI draw -i $GRAPH -c $PREFIX.lay -p $PREFIX.2D.png"

  PREFIX=$DIR_PARENT/$NAME.Ng
  sbatch -c 48 -p workers -J $NAME --wrap "hostname; $ODGI layout -i $GRAPH -N g -o $PREFIX.lay -T $PREFIX.lay.tsv -t 48 -P --temp-dir /scratch/; $ODGI draw -i $GRAPH -c $PREFIX.lay -p $PREFIX.2D.png"

  PREFIX=$DIR_PARENT/$NAME.x100.Ng
  sbatch -c 48 -p workers -J $NAME --wrap "hostname; $ODGI layout -i $GRAPH -N g -x 100 -o $PREFIX.lay -T $PREFIX.lay.tsv -t 48 -P --temp-dir /scratch/; $ODGI draw -i $GRAPH -c $PREFIX.lay -p $PREFIX.2D.png"
done

ls $DIR_GRAPHS/primates16.hsa6_p95.s5000.n16.k*/*.og | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)"
  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)

  PREFIX=$DIR_PARENT/$NAME.x1500.G20.I10000.l1000.Nh
  echo "$ODGI layout -i $GRAPH -x 1500 -G 20 -I 10000 -l 1000 -N h -o $PREFIX.lay -T $PREFIX.lay.tsv -t 96 -P --temp-dir /home/guarracino && $ODGI draw -i $GRAPH -c $PREFIX.lay -p $PREFIX.2D.png"
done
```

MHC:

```shell
cd $DIR_BASE

wget -c https://raw.githubusercontent.com/pangenome/odgi/master/test/chr6.HLA_genes.bed
bedtools merge -i chr6.HLA_genes.bed -d 10000000 | sed 's/^grch38/grch38#1/g' > chr6.MHC.bed

LOCUS=MHC
ls $DIR_GRAPHS/primates16.hsa6_p9*/*.final.og | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)/$LOCUS"
  mkdir -p $DIR_PARENT

  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)
  echo $NAME

  $ODGI extract -i $GRAPH -b chr6.MHC.bed -o - -O -t 48 -P | $ODGI sort -i - -o $DIR_PARENT/$NAME.$LOCUS.2.og -p gYs -x 300 -t 48 -P --temp-dir /scratch/
  
  $ODGI flip -i $DIR_PARENT/$NAME.$LOCUS.og -o $DIR_PARENT/$NAME.$LOCUS.flip.og -t 1
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.flip.og -o $DIR_PARENT/$NAME.$LOCUS.flip.z.png -z -c 64 -x 1500
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.flip.og -o $DIR_PARENT/$NAME.$LOCUS.flip.m.png -m -B Spectral:4 -c 64 -x 1500

  $ODGI sort -i $DIR_PARENT/$NAME.$LOCUS.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.og -p gYs -H <(odgi paths -i $DIR_PARENT/$NAME.$LOCUS.og -L | grep grch38) -x 300 -t 48 -P --temp-dir /scratch/
  $ODGI flip -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.og -t 1
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.z.png -z -c 64 -x 1500
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.m.png -m -B Spectral:4 -c 64 -x 1500
done
```

C4:

```shell
cd $DIR_BASE

wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes
wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ncbiRefSeq.gtf.gz
zgrep 'gene_id "C4A"\|gene_id "C4B"' hg38.ncbiRefSeq.gtf.gz |
  awk '$1 == "chr6"' | cut -f 1,4,5 |
  bedtools sort | bedtools merge -d 15000 | bedtools slop -l 10000 -r 20000 -g hg38.chrom.sizes |
  sed 's/chr6/grch38#1#chr6/g' > hg38.ncbiRefSeq.C4.bed

wget https://raw.githubusercontent.com/pangenome/odgi/master/test/chr6.C4.bed

LOCUS=C4
ls $DIR_GRAPHS/primates16.hsa6_p9*/*.final.og | while read GRAPH; do
  DIR_PARENT="$(dirname $GRAPH)/$LOCUS"
  mkdir -p $DIR_PARENT

  NAME=$(echo $GRAPH | rev | cut -f 2 -d '/' | rev)
  echo $NAME

  $ODGI extract -i $GRAPH -b hg38.ncbiRefSeq.C4.bed -o - -O -t 48 -P | $ODGI sort -i - -o $DIR_PARENT/$NAME.$LOCUS.og -p gYs -x 300 -t 48 -P --temp-dir /scratch/
  
  $ODGI flip -i $DIR_PARENT/$NAME.$LOCUS.og -o $DIR_PARENT/$NAME.$LOCUS.flip.og -t 1
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.flip.og -o $DIR_PARENT/$NAME.$LOCUS.flip.z.png -z -c 64 -x 1500
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.flip.og -o $DIR_PARENT/$NAME.$LOCUS.flip.m.png -m -B Spectral:4 -c 64 -x 1500

  $ODGI sort -i $DIR_PARENT/$NAME.$LOCUS.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.og -p gYs -H <(odgi paths -i $DIR_PARENT/$NAME.$LOCUS.og -L | grep grch38) -x 300 -t 48 -P --temp-dir /scratch/
  $ODGI flip -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.og -t 1
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.z.png -z -c 64 -x 1500
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.flip.m.png -m -B Spectral:4 -c 64 -x 1500

  $ODGI procbed -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.og -b <(sed 's/^grch38/grch38#1/g' chr6.C4.bed) > $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.C4.adj.bed
  $ODGI inject -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.og -b $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.C4.adj.bed -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.inject.og
  $ODGI flip -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.inject.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.inject.flip.og -t 1
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.inject.flip.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.inject.flip.z.png -z -c 64 -x 1500
  $ODGI viz -i $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.inject.flip.og -o $DIR_PARENT/$NAME.$LOCUS.sort-by-ref.inject.flip.m.png -m -B Spectral:4 -c 64 -x 1500
done
```




```shell
cd $DIR_GRAPHS/primates16.hsa6_p90.s5000.n16.k47.G700-900-1100.Pasm5.O0001_1/
cd $DIR_GRAPHS/primates16.hsa6_p90.s10000.n16.k79.G700-900-1100.Pasm5.O0001_1/

# Sort and add annotation
PREFIX=primates16.hsa6.fa.gz.bf3285f.c2fac19.9eace69.smooth.final
PREFIX=primates16.hsa6.fa.gz.c325321.c2fac19.9eace69.smooth.final
$ODGI sort -i $PREFIX.og -t $THREADS -o $PREFIX.Ygs.sort_by_ref.og -p Ygs -H <(echo chm13#1#chr6) -x 100 --temp-dir /scratch -P
$ODGI inject -i $PREFIX.Ygs.sort_by_ref.og -b /lizardfs/guarracino/pggb-paper/data/chr6.annotations.bed -o $PREFIX.Ygs.sort_by_ref.inject.og -t $THREADS -P

$ODGI paths -i $PREFIX.Ygs.sort_by_ref.inject.og -L | cut -f 1,2 -d '#' | sort | uniq > $PREFIX.Ygs.sort_by_ref.inject.prefixes.txt
$ODGI viz -i $PREFIX.Ygs.sort_by_ref.inject.og -o $PREFIX.Ygs.sort_by_ref.inject.merged_by_haplotype.z.png -x 1500 -y 500 -a 10 -z -I Consensus_ -M $PREFIX.Ygs.sort_by_ref.inject.prefixes.txt
$ODGI viz -i $PREFIX.Ygs.sort_by_ref.inject.og -o primates16.hsa6.fa.gz.bf3285f.e34d4cd.9eace69.smooth.final.Ygs.sort_by_ref.inject.merged_by_haplotype.du.png -x 1500 -y 500 -a 10 -du -I Consensus_ -M primates16.hsa6.fa.gz.bf3285f.e34d4cd.9eace69.smooth.final.Ygs.sort_by_ref.inject.prefixes.txt
```

## Notes

To remember that we are still doing this (both in PGGB and in GFA2EVALUATION) because of bugs in `vcfwave`:

```shell
# The TYPE info sometimes is wrong/missing
# There are variants without the ALT allele
bcftools annotate -x INFO/TYPE "$PATH_WAVED_VCF".gz  | awk '$5 != "."' | bgzip -@ THREADS -c > "$PATH_WAVED_FIXED_VCF".gz
```


## Minigraph Cactus

The first sample in `*.seqfile` is used to anchor the entire graph (https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/pangenome.md#sample-names).

Prepare the input:

```shell
cd $DIR_BASE

# TO RECONDIDER AND MAYBE TO REDO ATHALIANA82

sed '1d' $PATH_PANGENOMES_TSV | grep athaliana81 -v | cut -f 1 -d ' ' | sort | uniq | while read PATH_FASTA; do
  DIR_PANGENOME="$(dirname $PATH_FASTA)"

  >&2 echo $DIR_PANGENOME $PATH_FASTA

  cut -f 1 $PATH_FASTA.fai | sort -k 2,2n | cut -f 1,2 -d '#' | sort | uniq | while read ACC; do
    ACC2=$(echo $ACC | tr '#' '_')

    if [[ "$ACC" == "chm13#1" ]] || [[ "$ACC" == "grch38#1" ]]; then
        ACC3="${ACC%#*}"
    else
        ACC3="$(echo $ACC | tr '#' '.')"
    fi

    samtools faidx $PATH_FASTA $(cut -f 1 $PATH_FASTA.fai | grep $ACC) | bgzip -l 9 -@ 48 > $DIR_PANGENOME/$ACC2.fasta.gz
    samtools faidx $DIR_PANGENOME/$ACC2.fasta.gz
    echo $ACC3 $DIR_PANGENOME/$ACC2.fasta.gz | tr ' ' '\t'
  done > $DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
done

# Fix references
reorder_seqfile() {
    local seqfile=$1
    local ref=$2
    local temp_file="temp_seqfile"

    grep "$ref" "$seqfile" > "$temp_file"
    grep "$ref" "$seqfile" -v >> "$temp_file"
    mv "$temp_file" "$seqfile"
}
reorder_seqfile "$DIR_BASE/assemblies/primates/primates16.hsa6.seqfile" "grch38"
reorder_seqfile "$DIR_BASE/assemblies/tomato/tomato23.chr2.seqfile" "SL5"
reorder_seqfile "$DIR_BASE/assemblies/ecoli/ecoli50.seqfile" "GCA_000597845.1"
reorder_seqfile "$DIR_BASE/assemblies/ecoli/ecoli500.seqfile" "GCA_000597845.1"
# TODO reorder_seqfile "$DIR_BASE/assemblies/scerevisiae/scerevisiae142.seqfile" "SGDref"
# TODO reorder_seqfile "$DIR_BASE/assemblies/athaliana/athaliana82.seqfile" "GCA_028009825.1"


# On lambda01
mkdir -p $DIR_GRAPHS_MC

fmt="%C\n%Us user %Ss system %P cpu %es total %MKb max memory"
export PATH=/lizardfs/guarracino/tools/cactus-bin-v2.7.0/bin:$PATH
source /lizardfs/guarracino/tools/cactus-bin-v2.7.0/venv-cactus-v2.7.0/bin/activate

cd ~
sed '1d' $PATH_PANGENOMES_TSV | grep 'musculus' | cut -f 1 -d ' ' | sort | uniq | while read PATH_FASTA; do
  DIR_PANGENOME="$(dirname $PATH_FASTA)"
  PATH_SEQFILE=$DIR_PANGENOME/$(basename $PATH_FASTA .fa.gz).seqfile
  REF=$(head $PATH_SEQFILE -n 1 | cut -f 1)

  >&2 echo $PATH_SEQFILE $REF
  head $PATH_SEQFILE -n 1 | cut -f 1
  /usr/bin/time -f $fmt cactus-pangenome ./$(basename $PATH_FASTA .fa.gz)-js $PATH_SEQFILE --outDir $(basename $PATH_FASTA .fa.gz) --outName $(basename $PATH_FASTA .fa.gz) --reference $REF --vcf --giraffe --gfa --gbz --maxCores 64 --batchSystem single_machine --defaultCores 64 --vcfReference $REF --noSplit 2>&1 | tee $(basename $PATH_FASTA .fa.gz).MC.log; mv $(basename $PATH_FASTA .fa.gz)* $DIR_GRAPHS_MC
done









DIR_PANGENOME=/lizardfs/guarracino/pggb-paper/assemblies/athaliana
cut -f 1 $DIR_PANGENOME/athaliana82.fa.gz.fai| cut -f 1 -d '#' | sort | uniq | while read ACC; do echo $ACC; samtools faidx $DIR_PANGENOME/athaliana82.fa.gz $(cut -f 1 $DIR_PANGENOME/athaliana82.fa.gz.fai | grep $ACC) | bgzip -l 9 -@ 48 > $DIR_PANGENOME/$ACC.fasta.gz; samtools faidx $DIR_PANGENOME/$ACC.fasta.gz; done
rm Atha82.joblist
ls $DIR_PANGENOME/*fasta.gz | while read FASTA; do ACC=$(basename $FASTA .fasta.gz | cut -f 1,2 -d '_' | tr '.' '_'); echo "$ACC\t$FASTA" >> Atha82.joblist; done 

#!/bin/bash
fmt="%C\n%Us user %Ss system %P cpu %es total %MKb max memory"
export PATH=/lizardfs/guarracino/tools/cactus-bin-v2.7.0/bin:$PATH
source /lizardfs/guarracino/tools/cactus-bin-v2.7.0/venv-cactus-v2.7.0/bin/activate
/usr/bin/time -f $fmt cactus-pangenome ./js Atha82.joblist --outDir At82_MC --outName At82_MC --reference GCA_000001735_2 GCA_028009825_1 GCA_028009825_2 --vcf --giraffe --gfa --gbz --maxCores 224 --batchSystem single_machine --defaultCores 64 --vcfReference GCA_000001735_2 GCA_028009825_1 GCA_028009825_2
[2023-12-07T21:37:18+0000] [MainThread] [I] [toil.statsAndLogging] cactus-pangenome has finished after 89085.42339792382 seconds
2466690.39s user 69531.40s system 2846% cpu 89094.36s total 149417484Kb max memory

DIR_PANGENOME=/lizardfs/guarracino/pggb-paper/assemblies/primates
cut -f 1 $DIR_PANGENOME/primates16.hsa6.fa.gz.fai| cut -f 1 -d '#' | sort | uniq | while read ACC; do echo $ACC; samtools faidx $DIR_PANGENOME/primates16.hsa6.fa.gz $(cut -f 1 $DIR_PANGENOME/primates16.hsa6.fa.gz.fai | grep $ACC) | bgzip -l 9 -@ 48 > $DIR_PANGENOME/$ACC.hsa6.fasta.gz; samtools faidx $DIR_PANGENOME/$ACC.hsa6.fasta.gz; done
rm primates16.hsa6.joblist
ls $DIR_PANGENOME/*.hsa6.fasta.gz | while read FASTA; do ACC=$(basename $FASTA .fasta.gz | cut -f 1,2 -d '_' | tr '.' '_'); echo "$ACC\t$FASTA" >> primates16.hsa6.joblist; done 
[2023-12-07T08:28:49+0000] [MainThread] [I] [toil.statsAndLogging] cactus-pangenome has finished after 14534.32546418719 seconds
39138.20s user 7334.40s system 319% cpu 14536.33s total 57772840Kb max memory  

#!/bin/bash
fmt="%C\n%Us user %Ss system %P cpu %es total %MKb max memory"
export PATH=/lizardfs/guarracino/tools/cactus-bin-v2.7.0/bin:$PATH
source /lizardfs/guarracino/tools/cactus-bin-v2.7.0/venv-cactus-v2.7.0/bin/activate
/usr/bin/time -f $fmt cactus-pangenome ./js-primates16.hsa6 primates16.hsa6.joblist --outDir primates16.hsa6_MC --outName primates16.hsa6_MC --reference grch38_hsa6 --vcf --giraffe --gfa --gbz --maxCores 32 --batchSystem single_machine --defaultCores 8 --vcfReference grch38_hsa6
/usr/bin/time -f $fmt cactus-pangenome ./js-primates16.hsa6-no-split primates16.hsa6.joblist --outDir primates16.hsa6_MC-no-split --outName primates16.hsa6_MC-no-split --reference grch38_hsa6 --vcf --giraffe --gfa --gbz --maxCores 32 --batchSystem single_machine --defaultCores 8 --vcfReference grch38_hsa6 --noSplit
[2023-12-07T19:34:41+0000] [MainThread] [I] [toil.statsAndLogging] cactus-pangenome has finished after 14368.12507063197 seconds
36943.98s user 7407.71s system 308% cpu 14372.01s total 57775112Kb max memory
```

## PGGB with/without smoothxg

```shell
/home/guarracino/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9 -s /lizardfs/guarracino/pggb-paper/assemblies/hsapiens/hsapiens90.chr6.masked.fa.gz -p hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked.fa.gz.445f03b.alignments.wfmash.paf -k 47 -f 0 -g hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.gfa -B 10000000 -t 32 --temp-dir /scratch/ -P

/home/guarracino/tools/seqwish/bin/seqwish-f44b402f0c2e02988d431d9b2e5eba9727cf93a9 -s /lizardfs/guarracino/pggb-paper/assemblies/hsapiens/hsapiens90.chr6.masked.fa.gz -p hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked.fa.gz.445f03b.alignments.wfmash.paf -k 0 -f 0 -g hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1/hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.gfa -B 10000000 -t 32 --temp-dir /scratch/ -P


$ODGI extract -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.og -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -t 48 -P -O
$ODGI extract -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.gfa -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -t 48 -P -O
$ODGI extract -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.gfa -b ../../hg38.ncbiRefSeq.C4.bed -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -t 48 -P -O


$ODGI viz -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.unsorted.png -m -B Spectral:4
$ODGI sort -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -p gYs -o - -P -t 10 --temp-dir /scratch | $ODGI viz -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.png -m -B Spectral:4

$ODGI viz -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.unsorted.png -m -B Spectral:4
$ODGI sort -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -p gYs -o - -P -t 10 --temp-dir /scratch | $ODGI viz -i hsapiens90.chr6.masked.fa.gz.445f03b.e34d4cd.80ba7c5.smooth.final.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.png -m -B Spectral:4


$ODGI layout -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.lay --temp-dir /scratch -t 48 -P
$ODGI draw -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.og -c hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.lay -p hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k47.C4.2D.png

$ODGI layout -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -o hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.lay --temp-dir /scratch -t 48 -P
$ODGI draw -i hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.og -c hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.lay -p hsapiens90.chr6.masked_p98.s20000.n90.k47.G700-900-1100.Pasm5.O0001_1.seqwish.k0.C4.2D.png

ls *C4*.og | while read f; do echo $f; odgi stats -i $f -S; done

# quantify the runtime and memory too for odgi build/extract/sort?
# we would need to find a locus where -k 47 lead to underalignment and one where -k 0 lead to overalignment
```
