# Phylogeny

## All species

Variables:

```shell
ODGI=/home/guarracino/tools/odgi/bin/odgi-861b1c04f5622c5bb916a161c1abe812c213f1a5
DIR_GRAPHS=/lizardfs/guarracino/pggb-paper/graphs
DIR_PHYLOGENY=/lizardfs/guarracino/pggb-paper/phylogeny
PATH_PANGENOMES_TSV=/lizardfs/guarracino/pggb-paper/data/pangenomes.tsv
THREADS=48
```

Run `odgi paths`:

```shell
mkdir -p $DIR_PHYLOGENY

cd $DIR_GRAPHS
ls primates16.hsa6_p9*/*.og | while read GRAPH; do
  NAME=$(echo $GRAPH | cut -f 1 -d '/')
  echo $NAME

  $ODGI similarity -i $GRAPH -t $THREADS -d -D '#' -p 1> $DIR_PHYLOGENY/$NAME.dist.tsv
  $ODGI similarity -i $GRAPH -t $THREADS -d -D '#' -p 2 > $DIR_PHYLOGENY/$NAME.haplotype.dist.tsv
done


```shell
mkdir -p $DIR_PHYLOGENY
cd $DIR_PHYLOGENY
sed '1d' $PATH_PANGENOMES_TSV | grep 142 | while read f p s n k G POA O REF; do
  seq 3 | while read c; do
    echo $f $p $s $n $k $G $POA $O $REF $c;

    out=$(basename "$f" .fa.gz)_p$p.s$s.n$n.k$k.G$(echo $G | tr ',' '-').P$POA.O$(echo $O | sed 's/\.//g')_$c
    PATH_OG=$DIR_GRAPHS/$out/*.final.og
    PATH_OG=$(eval echo $PATH_OG) # force '*' expansion
    PREFIX=$(basename "eval $PATH_OG" .og)
    sbatch -c $THREADS -p workers -J $(basename "$f" .fa.gz) --wrap "hostname; $ODGI paths -i $PATH_OG -t $THREADS -d -D '#' > $DIR_PHYLOGENY/$PREFIX.$c.dist.tsv;"
  done
done
```

### Primates

```shell


/home/guarracino/tools/odgi/bin/odgi-odgi_paths_select_sep_pos paths -i /lizardfs/guarracino/pggb-paper/graphs/primates12.chr6_p90.s5000.n12.k79.G700-900-1100.Pasm5.O0001_1/primates12.chr6.fa.gz.a130aa2.c2fac19.0bfc78d.smooth.final.sed.gfa -t 48 -d -D '#' -p 0 -P > /lizardfs/guarracino/pggb-paper/graphs/primates12.chr6_p90.s5000.n12.k79.G700-900-1100.Pasm5.O0001_1/primates12.chr6.fa.gz.a130aa2.c2fac19.0bfc78d.smooth.final.1.dist.samples.tsv

/home/guarracino/tools/odgi/bin/odgi-odgi_paths_select_sep_pos paths -i /lizardfs/guarracino/pggb-paper/graphs/primates12.chr6_p90.s5000.n12.k79.G700-900-1100.Pasm5.O0001_1/primates12.chr6.fa.gz.a130aa2.c2fac19.0bfc78d.smooth.final.sed.gfa -t 48 -d -D '#' -p 1 -P > /lizardfs/guarracino/pggb-paper/graphs/primates12.chr6_p90.s5000.n12.k79.G700-900-1100.Pasm5.O0001_1/primates12.chr6.fa.gz.a130aa2.c2fac19.0bfc78d.smooth.final.1.dist.haplotype.tsv



grep '^P' primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.gfa | cut -f 2 | cut -f 1,2 -d '#' | sort | uniq > primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.prefixes.txt

$ODGI viz -i primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.og -M primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.prefixes.txt -du -o primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.du.haplotypes.png



sed 's/#chr/#1#chr/g' /lizardfs/guarracino/pggb-paper/graphs/primates14.chr6_p90.s5000.n14.k79.G700-900-1100.Pasm5.O0001_1/primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.gfa > /lizardfs/guarracino/pggb-paper/graphs/primates14.chr6_p90.s5000.n14.k79.G700-900-1100.Pasm5.O0001_1/primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.gfa

/home/guarracino/tools/odgi/bin/odgi-odgi_paths_select_sep_pos paths -i /lizardfs/guarracino/pggb-paper/graphs/primates14.chr6_p90.s5000.n14.k79.G700-900-1100.Pasm5.O0001_1/primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.gfa -t 48 -d -D '#' -p 0 -P > /lizardfs/guarracino/pggb-paper/graphs/primates14.chr6_p90.s5000.n14.k79.G700-900-1100.Pasm5.O0001_1/primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.1.dist.samples.tsv

/home/guarracino/tools/odgi/bin/odgi-odgi_paths_select_sep_pos paths -i /lizardfs/guarracino/pggb-paper/graphs/primates14.chr6_p90.s5000.n14.k79.G700-900-1100.Pasm5.O0001_1/primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.sed.gfa -t 48 -d -D '#' -p 1 -P > /lizardfs/guarracino/pggb-paper/graphs/primates14.chr6_p90.s5000.n14.k79.G700-900-1100.Pasm5.O0001_1/primates14.chr6.fa.gz.667b9b6.c2fac19.ee137be.smooth.final.1.dist.haplotype.tsv
```

Make phylogenetic trees:

```shell
# To finish
```
