# Phylogeny

Variables:

```shell
DIR_BASE=/lizardfs/guarracino/pggb-paper
DIR_GRAPHS=/lizardfs/guarracino/pggb-paper/graphs
DIR_PHYLOGENY=/lizardfs/guarracino/pggb-paper/phylogeny

ODGI=/home/guarracino/tools/odgi/bin/odgi-861b1c04f5622c5bb916a161c1abe812c213f1a5
PATH_PANGENOMES_TSV=/lizardfs/guarracino/pggb-paper/data/pangenomes.tsv

THREADS=48
```

## Primates

Run `odgi paths`:

```shell
mkdir -p $DIR_PHYLOGENY

cd $DIR_GRAPHS
ls primates16.hsa6_p9*/*final.og | while read GRAPH; do
  NAME=$(echo $GRAPH | cut -f 1 -d '/')
  echo $NAME

  $ODGI similarity -i $GRAPH -d -D '#' -p 1 -t $THREADS > $DIR_PHYLOGENY/$NAME.dist.tsv
  $ODGI similarity -i $GRAPH -d -D '#' -p 2 -t $THREADS > $DIR_PHYLOGENY/$NAME.haplotype.dist.tsv
done

(ls primates16.hsa6_p9*/C4/*C4.og; ls primates16.hsa6_p9*/MHC/*MHC.og) | while read GRAPH; do
  NAME=$(basename $GRAPH .og)
  echo $NAME

  $ODGI similarity -i $GRAPH -d -D '#' -p 1 -t $THREADS > $DIR_PHYLOGENY/$NAME.dist.tsv
  $ODGI similarity -i $GRAPH -d -D '#' -p 2 -t $THREADS > $DIR_PHYLOGENY/$NAME.haplotype.dist.tsv
done
```

Make phylogenetic trees:

```shell
ls $DIR_PHYLOGENY/*.dist.tsv | while read f; do
  NAME=$(basename $f .dist.tsv);
  echo $NAME;

  Rscript $DIR_BASE/scripts/trees.R $f $NAME $DIR_PHYLOGENY/$NAME.dist.pdf
done
```
